
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Exploratory Data Analysis for Metadata &#8212; scRNA Sequencing Analysis</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'EDA_metadata';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Exploratory Data Analysis for the sparse matrix" href="EDA_sparse_matrix.html" />
    <link rel="prev" title="Important Concepts and Metrics Used for scRNA-seq Analysis" href="glossary.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo_uninorte.png" class="logo__image only-light" alt="scRNA Sequencing Analysis - Home"/>
    <script>document.write(`<img src="_static/logo_uninorte.png" class="logo__image only-dark" alt="scRNA Sequencing Analysis - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Final Project - Deep Learning
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction to Single-Cell RNA Sequencing</a></li>

<li class="toctree-l1"><a class="reference internal" href="glossary.html">Important Concepts and Metrics Used for scRNA-seq Analysis</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Exploratory Data Analysis for Metadata</a></li>

<li class="toctree-l1"><a class="reference internal" href="EDA_sparse_matrix.html">Exploratory Data Analysis for the sparse matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="Datos_celulares_y_geneticos.html">Exploratory Data Analysis for Glioblastoma data</a></li>
<li class="toctree-l1"><a class="reference internal" href="Agrupaciones_y_clusters.html">Exploratory Data Analysis for clusterization and gene expressions</a></li>
<li class="toctree-l1"><a class="reference internal" href="EDA_test.html">Exploratory Data Analysis for test data</a></li>

<li class="toctree-l1"><a class="reference internal" href="preprocessing.html">Preprocessing of Raw Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="PCA.html">Principal Component Analysis - Dimensionality reduction</a></li>

<li class="toctree-l1"><a class="reference internal" href="Benchmark_Models.html">Benchmark Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="spaco.html">Spatial Component Analysis Based on Latent Representations (SpaCO)</a></li>

</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/henrysaenz/eda_rna" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/henrysaenz/eda_rna/issues/new?title=Issue%20on%20page%20%2FEDA_metadata.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/EDA_metadata.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Exploratory Data Analysis for Metadata</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Exploratory Data Analysis for Metadata</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#import-data">Import data</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#analyze-and-describe-data">Analyze and Describe data</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-for-cell-type-and-proportion">Count for cell type and proportion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-and-proportion-of-subtissue">Count and proportion of subtissue</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#missing-data">Missing data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#statistical-summary">Statistical summary</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analysis-for-each-sex">Analysis For each sex</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-and-proportion-of-tissues-type-by-sex">Count and proportion of tissues type by sex</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cell-count-by-type">Cell count by type</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cell-count-by-tissue-type">Cell count by tissue type</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#heatmap-of-gene-expression">Heatmap of gene expression</a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="exploratory-data-analysis-for-metadata">
<h1>Exploratory Data Analysis for Metadata<a class="headerlink" href="#exploratory-data-analysis-for-metadata" title="Link to this heading">#</a></h1>
<section id="import-data">
<h2>Import data<a class="headerlink" href="#import-data" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">IPython.core.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">display</span><span class="p">,</span> <span class="n">HTML</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C:\Users\henry\AppData\Local\Temp\ipykernel_17604\1582991288.py:6: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
  from IPython.core.display import display, HTML
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_path</span> <span class="o">=</span> <span class="s2">&quot;C:/Users/henry/Desktop/python/jbook_rna/datos/Single-cell RNASeq data from Mouse Brain&quot;</span>
<span class="n">data_path2</span> <span class="o">=</span> <span class="s2">&quot;C:/Users/henry/Desktop/python/jbook_rna/datos/Theory - Intro to single-cell RNAseq Images&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">brain_counts_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="s2">&quot;brain_counts.csv&quot;</span><span class="p">)</span>
<span class="n">brain_metadata_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="s2">&quot;brain_metadata.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_counts</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">brain_counts_file</span><span class="p">)</span>
<span class="n">df_metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">brain_metadata_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="analyze-and-describe-data">
<h1>Analyze and Describe data<a class="headerlink" href="#analyze-and-describe-data" title="Link to this heading">#</a></h1>
<p>Metadata is “data about data.” In this dataset, it provides descriptive, structural, and administrative information about a mice laboratory gene and RNA experiment, where it was collected cellular and RNA decomposition information from 3401 male and female population.</p>
<p>The dimension of the data is 3401 observation for 6 different variables. This information belongs to the decsription of the the brain extracted mices cells</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Dimension of DataFrame:&quot;</span><span class="p">,</span> <span class="n">df_metadata</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Dimension of DataFrame: (3401, 6)
</pre></div>
</div>
</div>
</div>
<p>All columns are of type object (string/text), meaning they contain categorical data. Below is an overview of each column:</p>
<ul class="simple">
<li><p><strong>cell</strong>: Identifies individual cells in the dataset. Each row corresponds to a unique cell.</p></li>
<li><p><strong>cell_ontology_class</strong>: Specifies the ontology classification of each cell, likely indicating its biological type or function.</p></li>
<li><p><strong>subtissue</strong>: Describes the specific tissue or sub-region from which the cell was extracted.</p></li>
<li><p><strong>mouse.sex:</strong> Indicates the sex of the mouse (e.g., “male” or “female”).</p></li>
<li><p><strong><a class="reference external" href="http://mouse.id">mouse.id</a>:</strong> A unique identifier for each mouse in the study. Multiple cells may belong to the same mouse.</p></li>
<li><p><strong>plate.barcode:</strong> A barcode identifier for the plate used in the experiment, which can help track batch effects or technical variations.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;📊 Information of DataFrame:&quot;</span><span class="p">)</span>
<span class="n">df_metadata</span><span class="o">.</span><span class="n">info</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>📊 Information of DataFrame:
&lt;class &#39;pandas.core.frame.DataFrame&#39;&gt;
RangeIndex: 3401 entries, 0 to 3400
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   cell                 3401 non-null   object
 1   cell_ontology_class  3401 non-null   object
 2   subtissue            3401 non-null   object
 3   mouse.sex            3401 non-null   object
 4   mouse.id             3401 non-null   object
 5   plate.barcode        3401 non-null   object
dtypes: object(6)
memory usage: 159.5+ KB
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p><strong>cell_ontology_class</strong>: Cell classification with <strong>7 distinct types</strong>, with <strong>oligodendrocytes</strong> being the most common (<strong>1,574 occurrences</strong>).</p></li>
<li><p><strong>subtissue</strong>: The brain sub-region where the cell was extracted (<strong>4 unique regions</strong>), with the <strong>Cortex</strong> as the most frequent (<strong>1,149 occurrences</strong>).</p></li>
<li><p><strong>mouse.sex</strong>: Mouse sex (<strong>2 categories</strong>: Male/Female), with <strong>males being the majority</strong> (<strong>2,694 occurrences</strong>).</p></li>
<li><p><strong><a class="reference external" href="http://mouse.id">mouse.id</a></strong>: Unique mouse identifier (<strong>7 distinct mice</strong>), with <strong>mouse “3_10_M” being the most frequent</strong> (<strong>980 occurrences</strong>).</p></li>
<li><p><strong>plate.barcode</strong>: Experimental plate barcode (<strong>41 unique plates</strong>), with <strong>“MAA000560” used most often</strong> (<strong>28 occurrences</strong>).</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">📈 Description of numerical variables:&quot;</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">HTML</span><span class="p">(</span><span class="n">df_metadata</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span><span class="o">.</span><span class="n">to_html</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>📈 Description of numerical variables:
</pre></div>
</div>
<div class="output text_html"><table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cell</th>
      <th>cell_ontology_class</th>
      <th>subtissue</th>
      <th>mouse.sex</th>
      <th>mouse.id</th>
      <th>plate.barcode</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>3401</td>
      <td>3401</td>
      <td>3401</td>
      <td>3401</td>
      <td>3401</td>
      <td>3401</td>
    </tr>
    <tr>
      <th>unique</th>
      <td>3401</td>
      <td>7</td>
      <td>4</td>
      <td>2</td>
      <td>7</td>
      <td>41</td>
    </tr>
    <tr>
      <th>top</th>
      <td>A1.B003290.3_38_F.1.1</td>
      <td>oligodendrocyte</td>
      <td>Cortex</td>
      <td>M</td>
      <td>3_10_M</td>
      <td>MAA000560</td>
    </tr>
    <tr>
      <th>freq</th>
      <td>1</td>
      <td>1574</td>
      <td>1149</td>
      <td>2694</td>
      <td>980</td>
      <td>287</td>
    </tr>
  </tbody>
</table></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">🔹 Firts 10 rows of the DataFrame:&quot;</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">HTML</span><span class="p">(</span><span class="n">df_metadata</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span><span class="o">.</span><span class="n">to_html</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>🔹 Firts 10 rows of the DataFrame:
</pre></div>
</div>
<div class="output text_html"><table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>cell</th>
      <th>cell_ontology_class</th>
      <th>subtissue</th>
      <th>mouse.sex</th>
      <th>mouse.id</th>
      <th>plate.barcode</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>A1.B003290.3_38_F.1.1</td>
      <td>astrocyte</td>
      <td>Striatum</td>
      <td>F</td>
      <td>3_38_F</td>
      <td>B003290</td>
    </tr>
    <tr>
      <td>A1.B003728.3_56_F.1.1</td>
      <td>astrocyte</td>
      <td>Striatum</td>
      <td>F</td>
      <td>3_56_F</td>
      <td>B003728</td>
    </tr>
    <tr>
      <td>A1.MAA000560.3_10_M.1.1</td>
      <td>oligodendrocyte</td>
      <td>Cortex</td>
      <td>M</td>
      <td>3_10_M</td>
      <td>MAA000560</td>
    </tr>
    <tr>
      <td>A1.MAA000564.3_10_M.1.1</td>
      <td>endothelial cell</td>
      <td>Striatum</td>
      <td>M</td>
      <td>3_10_M</td>
      <td>MAA000564</td>
    </tr>
    <tr>
      <td>A1.MAA000923.3_9_M.1.1</td>
      <td>astrocyte</td>
      <td>Hippocampus</td>
      <td>M</td>
      <td>3_9_M</td>
      <td>MAA000923</td>
    </tr>
    <tr>
      <td>A1.MAA000930.3_8_M.1.1</td>
      <td>oligodendrocyte</td>
      <td>Cortex</td>
      <td>M</td>
      <td>3_8_M</td>
      <td>MAA000930</td>
    </tr>
    <tr>
      <td>A1.MAA000932.3_11_M.1.1</td>
      <td>endothelial cell</td>
      <td>Hippocampus</td>
      <td>M</td>
      <td>3_11_M</td>
      <td>MAA000932</td>
    </tr>
    <tr>
      <td>A1.MAA000940.3_8_M.1.1</td>
      <td>neuron</td>
      <td>Striatum</td>
      <td>M</td>
      <td>3_8_M</td>
      <td>MAA000940</td>
    </tr>
    <tr>
      <td>A1.MAA001854.3_38_F.1.1</td>
      <td>oligodendrocyte</td>
      <td>Cerebellum</td>
      <td>F</td>
      <td>3_38_F</td>
      <td>MAA001854</td>
    </tr>
    <tr>
      <td>A1.MAA001894.3_39_F.1.1</td>
      <td>endothelial cell</td>
      <td>Cortex</td>
      <td>F</td>
      <td>3_39_F</td>
      <td>MAA001894</td>
    </tr>
  </tbody>
</table></div></div>
</div>
<section id="count-for-cell-type-and-proportion">
<h2>Count for cell type and proportion<a class="headerlink" href="#count-for-cell-type-and-proportion" title="Link to this heading">#</a></h2>
<p>The following chart presents the distribution of cell types in a dataset of mouse brain cells. The left plot shows the absolute count of each cell type, while the right plot displays their proportional distribution as a percentage.</p>
<p><strong>Key observations:</strong></p>
<ul class="simple">
<li><p><em>Oligodendrocytes</em> are the most abundant cell type, making up the largest portion of the dataset.</p></li>
<li><p><em>Endothelial</em> cells and astrocytes are also common, though in lower numbers.</p></li>
<li><p><em>Neurons</em>, <em>oligodendrocyte</em> <em>precursor</em> <em>cells</em>, and <em>brain pericytes</em> are present in moderate amounts.</p></li>
<li><p>Bergmann glial cells appear in the smallest proportion.</p></li>
</ul>
<p>The proportional chart highlights the dominance of oligodendrocytes, which account for over 40% of the dataset, while other cell types have significantly lower representation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_type_counts</span> <span class="o">=</span> <span class="n">df_metadata</span><span class="p">[</span><span class="s2">&quot;cell_ontology_class&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
<span class="n">cell_type_percent</span> <span class="o">=</span> <span class="p">(</span><span class="n">cell_type_counts</span> <span class="o">/</span> <span class="n">cell_type_counts</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span> <span class="o">*</span> <span class="mi">100</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">cell_type_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">cell_type_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Count for cell type (Quantity)&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Cell type&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Quantity&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>

<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">cell_type_percent</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">cell_type_percent</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Cell Type Count (Proportion)&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Cell type&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Percentage (%)&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C:\Users\henry\AppData\Local\Temp\ipykernel_17604\3752527540.py:6: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(x=cell_type_counts.index, y=cell_type_counts.values, palette=&quot;coolwarm&quot;, ax=axes[0])
C:\Users\henry\AppData\Local\Temp\ipykernel_17604\3752527540.py:12: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(x=cell_type_percent.index, y=cell_type_percent.values, palette=&quot;coolwarm&quot;, ax=axes[1])
</pre></div>
</div>
<img alt="_images/f9f3773d4dd8bfe1a7984e6eb82fded85cdb36cd82d1fbee23ad1842bb873591.png" src="_images/f9f3773d4dd8bfe1a7984e6eb82fded85cdb36cd82d1fbee23ad1842bb873591.png" />
</div>
</div>
</section>
<section id="count-and-proportion-of-subtissue">
<h2>Count and proportion of subtissue<a class="headerlink" href="#count-and-proportion-of-subtissue" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p>The left Chart “Subtissue Distribution (Quantity)” displays the absolute count of cells in different brain subtissues. The Cortex has the highest count, followed by the Hippocampus, Striatum, and Cerebellum, which has the lowest count.</p></li>
<li><p>The right chart “Proportion of Subtissue” represents the proportion of each subtissue relative to the total dataset. The Cortex accounts for the largest percentage (33.8%), followed by the Hippocampus (28.7%), Striatum (21.3%), and Cerebellum (16.3%).</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">subtissue_type_counts</span> <span class="o">=</span> <span class="n">df_metadata</span><span class="p">[</span><span class="s1">&#39;subtissue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
<span class="n">subtissue_type_percent</span> <span class="o">=</span> <span class="p">(</span><span class="n">subtissue_type_counts</span> <span class="o">/</span> <span class="n">subtissue_type_counts</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span> <span class="o">*</span> <span class="mi">100</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>


<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">subtissue_type_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">subtissue_type_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Subtissue Distribution (Quantity)&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Subtissue&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Quantity&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>

<span class="n">colors</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">subtissue_type_counts</span><span class="p">))</span>  <span class="c1"># Colores</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">subtissue_type_percent</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">subtissue_type_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">autopct</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%1.1f%%</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">colors</span><span class="p">,</span> <span class="n">startangle</span><span class="o">=</span><span class="mi">140</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Proportion of Subtissue&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C:\Users\henry\AppData\Local\Temp\ipykernel_17604\2567555667.py:8: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(x=subtissue_type_counts.index, y=subtissue_type_counts.values, palette=&quot;coolwarm&quot;, ax=axes[0])
</pre></div>
</div>
<img alt="_images/548e30a1b04e0f70f60da4250657fdf4a1c922e7966ad144fe548f7d2c2f7702.png" src="_images/548e30a1b04e0f70f60da4250657fdf4a1c922e7966ad144fe548f7d2c2f7702.png" />
</div>
</div>
</section>
<section id="missing-data">
<h2>Missing data<a class="headerlink" href="#missing-data" title="Link to this heading">#</a></h2>
<p>The missing data analysis shows that there is no missing data in any of the columns. Each column; cell, cell_ontology_class, subtissue, mouse.sex, <a class="reference external" href="http://mouse.id">mouse.id</a>, and plate.barcode—has complete values, meaning there are no null or empty entries.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To check missing data</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">🔍 Missing data in Metadata:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_metadata</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>🔍 Missing data in Metadata:
cell                   0
cell_ontology_class    0
subtissue              0
mouse.sex              0
mouse.id               0
plate.barcode          0
dtype: int64
</pre></div>
</div>
</div>
</div>
</section>
<section id="statistical-summary">
<h2>Statistical summary<a class="headerlink" href="#statistical-summary" title="Link to this heading">#</a></h2>
<p><strong>Key Observations:</strong></p>
<ul class="simple">
<li><p><em><strong>Cell Ontology Class:</strong></em> There are 7 distinct cell types, with oligodendrocytes being the most frequent (1,574 occurrences).</p></li>
<li><p><em><strong>Subtissue Distribution:</strong></em> Cells are found in 4 different brain regions, with the Cortex being the most represented (1,149 occurrences).</p></li>
<li><p><em><strong>Sex Distribution:</strong></em> The dataset contains both male and female samples, but males (2,694) are more prevalent.</p></li>
<li><p><em><strong>Mouse ID:</strong></em> There are 7 unique mouse IDs, with mouse 3_10_M appearing most frequently (980 occurrences).</p></li>
<li><p><em><strong>Plate Barcode:</strong></em> A total of 41 different plate barcodes were used, with MAA000560 being the most common (287 occurrences).</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Statistical summary </span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">📈 Statistical summary:&quot;</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">HTML</span><span class="p">(</span><span class="n">df_metadata</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span><span class="o">.</span><span class="n">to_html</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>📈 Statistical summary:
</pre></div>
</div>
<div class="output text_html"><table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cell</th>
      <th>cell_ontology_class</th>
      <th>subtissue</th>
      <th>mouse.sex</th>
      <th>mouse.id</th>
      <th>plate.barcode</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>3401</td>
      <td>3401</td>
      <td>3401</td>
      <td>3401</td>
      <td>3401</td>
      <td>3401</td>
    </tr>
    <tr>
      <th>unique</th>
      <td>3401</td>
      <td>7</td>
      <td>4</td>
      <td>2</td>
      <td>7</td>
      <td>41</td>
    </tr>
    <tr>
      <th>top</th>
      <td>A1.B003290.3_38_F.1.1</td>
      <td>oligodendrocyte</td>
      <td>Cortex</td>
      <td>M</td>
      <td>3_10_M</td>
      <td>MAA000560</td>
    </tr>
    <tr>
      <th>freq</th>
      <td>1</td>
      <td>1574</td>
      <td>1149</td>
      <td>2694</td>
      <td>980</td>
      <td>287</td>
    </tr>
  </tbody>
</table></div></div>
</div>
<p>There are also zero values in the gene expression matrix wich means that the matrix has full values.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Null values</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">🔍 Null Values in the gene expresssion matrix:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_counts</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>🔍 Null Values in the gene expresssion matrix:
0
</pre></div>
</div>
</div>
</div>
</section>
<section id="analysis-for-each-sex">
<h2>Analysis For each sex<a class="headerlink" href="#analysis-for-each-sex" title="Link to this heading">#</a></h2>
<p>The metadata information of the cells analyzed contains 2694 registers that belongs to male mice and 707 that belongs to female.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_metadata_m</span> <span class="o">=</span> <span class="n">df_metadata</span><span class="p">[</span><span class="n">df_metadata</span><span class="p">[</span><span class="s1">&#39;mouse.sex&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;M&#39;</span><span class="p">]</span>
<span class="n">df_metadata_f</span> <span class="o">=</span> <span class="n">df_metadata</span><span class="p">[</span><span class="n">df_metadata</span><span class="p">[</span><span class="s1">&#39;mouse.sex&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;F&#39;</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1"> We filter by sex of the mouse and we get that: &#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;🐭 Male: </span><span class="si">{</span><span class="n">df_metadata_m</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;🐭 Female: </span><span class="si">{</span><span class="n">df_metadata_f</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span> We filter by sex of the mouse and we get that: 
🐭 Male: 2694
🐭 Female: 707
</pre></div>
</div>
</div>
</div>
<p>The mice proportion in the dataset consist in 20.8% for female and 79,8% for male, wich is the greatest part.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sex_counts</span> <span class="o">=</span> <span class="n">df_metadata</span><span class="p">[</span><span class="s1">&#39;mouse.sex&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">sex_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">sex_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;pastel&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Mice count per sex&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Sex&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Quantity&quot;</span><span class="p">)</span>

<span class="c1"># Pie chart</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">sex_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">sex_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">autopct</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%1.1f%%</span><span class="s1">&#39;</span><span class="p">,</span> 
            <span class="n">colors</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;pastel&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sex_counts</span><span class="p">)),</span> <span class="n">startangle</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Sex proportion of mice&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C:\Users\henry\AppData\Local\Temp\ipykernel_17604\798154539.py:5: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(x=sex_counts.index, y=sex_counts.values, palette=&quot;pastel&quot;, ax=axes[0])
</pre></div>
</div>
<img alt="_images/cd05c7482b619d6f4c15bf62da5112c7133cb8328e4b793a8b5646be256b9c58.png" src="_images/cd05c7482b619d6f4c15bf62da5112c7133cb8328e4b793a8b5646be256b9c58.png" />
</div>
</div>
</section>
<section id="count-and-proportion-of-tissues-type-by-sex">
<h2>Count and proportion of tissues type by sex<a class="headerlink" href="#count-and-proportion-of-tissues-type-by-sex" title="Link to this heading">#</a></h2>
<p>The male group has a significantly larger sample size compared to the female group, which is evident from the higher counts across all tissues.</p>
<ul class="simple">
<li><p>Males show a more skewed distribution, with Cortex having the highest sample count and the other tissues decreasing progressively.</p></li>
<li><p>Females have a more balanced distribution among Cortex, Hippocampus, and Striatum, while Cerebellum remains the least represented in both groups.</p></li>
</ul>
<p>While absolute counts are different, the relative proportions (percentages) provide a normalized view of tissue representation:</p>
<ul class="simple">
<li><p>Males have a higher proportion of <strong>Cortex</strong> samples (35.0%) compared to females (29.3%).</p></li>
<li><p><strong>Striatum</strong> is more represented in females (27.0%) than in males (19.7%).</p></li>
<li><p><strong>Hippocampus</strong> is nearly identical in both groups (~28%).</p></li>
<li><p><strong>Cerebellum</strong> has the lowest proportion in both, with a slight difference between males (16.5%) and females (15.3%).</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tissue_counts_m</span> <span class="o">=</span> <span class="n">df_metadata</span><span class="p">[</span><span class="n">df_metadata</span><span class="p">[</span><span class="s1">&#39;mouse.sex&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;M&#39;</span><span class="p">][</span><span class="s1">&#39;subtissue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
<span class="n">tissue_counts_f</span> <span class="o">=</span> <span class="n">df_metadata</span><span class="p">[</span><span class="n">df_metadata</span><span class="p">[</span><span class="s1">&#39;mouse.sex&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;F&#39;</span><span class="p">][</span><span class="s1">&#39;subtissue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">tissue_counts_m</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">tissue_counts_m</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Tissue count for male&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Tissue&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Quantity&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>

<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">tissue_counts_f</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">tissue_counts_f</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Tissue count for female&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Tissue&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>

<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">tissue_counts_m</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">tissue_counts_m</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">autopct</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%1.1f%%</span><span class="s1">&#39;</span><span class="p">,</span> 
               <span class="n">colors</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">tissue_counts_m</span><span class="p">)))</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Tissue proportion for male&quot;</span><span class="p">)</span>

<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">tissue_counts_f</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">tissue_counts_f</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">autopct</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%1.1f%%</span><span class="s1">&#39;</span><span class="p">,</span> 
               <span class="n">colors</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">tissue_counts_f</span><span class="p">)))</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Tissue proportion for female&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C:\Users\henry\AppData\Local\Temp\ipykernel_17604\2428431691.py:6: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(x=tissue_counts_m.index, y=tissue_counts_m.values, palette=&quot;coolwarm&quot;, ax=axes[0, 0])
C:\Users\henry\AppData\Local\Temp\ipykernel_17604\2428431691.py:12: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(x=tissue_counts_f.index, y=tissue_counts_f.values, palette=&quot;coolwarm&quot;, ax=axes[0, 1])
</pre></div>
</div>
<img alt="_images/66591e4bc1b9159068793c12a1c123908e37045b65fbbcdc644364a1b5f5c747.png" src="_images/66591e4bc1b9159068793c12a1c123908e37045b65fbbcdc644364a1b5f5c747.png" />
</div>
</div>
</section>
<section id="cell-count-by-type">
<h2>Cell count by type<a class="headerlink" href="#cell-count-by-type" title="Link to this heading">#</a></h2>
<p>These information is divided into counts per cell type as followed:</p>
<ul class="simple">
<li><p><em>Oligodendrocytes</em> are the most numerous, with approximately 1,600 cells, making them the dominant cell type.</p></li>
<li><p><em>Endothelial cells</em> are the second most common, with about 700 cells, representing a significant portion of the dataset.</p></li>
<li><p><em>Astrocytes</em> account for roughly 430 cells, while neurons are less frequent, with around 280 cells.</p></li>
<li><p>The remaining cell <em>types—oligodendrocyte</em> <em>precursor cells</em> (~200), <em>brain pericytes</em> (~150), and <em>Bergmann glial cells</em> (~40)—are comparatively rare.</p></li>
</ul>
<p><strong>Statistical Insights</strong></p>
<ul class="simple">
<li><p><em><strong>Skewed Distribution:</strong></em> The dataset is highly unbalanced, with oligodendrocytes being nearly twice as common as the next most frequent cell type (endothelial cells).</p></li>
<li><p><em><strong>Low Representation:</strong></em> Some cell types, such as Bergmann glial cells (1.2%) and brain pericytes (4.6%), have very few samples, which may impact statistical analyses if these groups are too small to generalize results.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_counts</span> <span class="o">=</span> <span class="n">df_metadata</span><span class="p">[</span><span class="s1">&#39;cell_ontology_class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">cell_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">cell_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Cell count by type&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Cell type&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Quantity&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>

<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">cell_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">cell_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">autopct</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%1.1f%%</span><span class="s1">&#39;</span><span class="p">,</span> 
            <span class="n">colors</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">cell_counts</span><span class="p">)))</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Cell type proportion&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C:\Users\henry\AppData\Local\Temp\ipykernel_17604\2006688091.py:7: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(x=cell_counts.index, y=cell_counts.values, palette=&quot;coolwarm&quot;, ax=axes[0])
</pre></div>
</div>
<img alt="_images/e997e2fb1e177fff39fbb03636713a3980b238c4b482c06eb9744b20b1c3bc0a.png" src="_images/e997e2fb1e177fff39fbb03636713a3980b238c4b482c06eb9744b20b1c3bc0a.png" />
</div>
</div>
</section>
<section id="cell-count-by-tissue-type">
<h2>Cell count by tissue type<a class="headerlink" href="#cell-count-by-tissue-type" title="Link to this heading">#</a></h2>
<p><strong>Oligodendrocytes</strong></p>
<ul class="simple">
<li><p>Most abundant in Cortex (~580 cells), followed by Hippocampus (~470) and Striatum (~380).</p></li>
<li><p>Least common in Cerebellum (~170).</p></li>
</ul>
<p><strong>Endothelial Cells:</strong></p>
<ul class="simple">
<li><p>Highest in Hippocampus (~270), followed by Cerebellum (~190).</p></li>
<li><p>Lower presence in Cortex (~140) and Striatum (~110).</p></li>
</ul>
<p><strong>Astrocytes</strong>:</p>
<ul class="simple">
<li><p>Predominantly found in Cortex (~260).</p></li>
<li><p>Lower numbers in Hippocampus (~90), Striatum (~80), and Cerebellum (~10).</p></li>
</ul>
<p><strong>Neurons</strong>:</p>
<ul class="simple">
<li><p>More evenly distributed, with Striatum (~110) and Cerebellum (~100) having slightly higher counts.</p></li>
<li><p>Cortex and Hippocampus have fewer neurons (~50 and ~30, respectively).</p></li>
</ul>
<p><strong>Oligodendrocyte Precursor Cells:</strong></p>
<ul class="simple">
<li><p>Most present in Hippocampus (~100), with Cortex (~90) and Striatum (~50) having moderate counts.</p></li>
<li><p>Least frequent in Cerebellum (~20).</p></li>
</ul>
<p><strong>Brain Pericytes:</strong></p>
<ul class="simple">
<li><p>Found in all tissues in small amounts, with slightly more in Cortex (~80) and Cerebellum (~70).</p></li>
</ul>
<p><strong>Bergmann Glial Cells:</strong></p>
<ul class="simple">
<li><p>Almost exclusive to Cerebellum (~60), with negligible counts in other tissues.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tissue_by_cell</span> <span class="o">=</span> <span class="n">df_metadata</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;cell_ontology_class&#39;</span><span class="p">,</span> <span class="s1">&#39;subtissue&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;count&#39;</span><span class="p">)</span>
<span class="n">tissue_by_cell</span> <span class="o">=</span> <span class="n">tissue_by_cell</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="n">tissue_by_cell</span><span class="p">,</span> 
    <span class="n">x</span><span class="o">=</span><span class="s1">&#39;cell_ontology_class&#39;</span><span class="p">,</span> 
    <span class="n">y</span><span class="o">=</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> 
    <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;subtissue&#39;</span><span class="p">,</span> 
    <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;coolwarm&#39;</span><span class="p">,</span> 
    <span class="n">order</span><span class="o">=</span><span class="n">tissue_by_cell</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;cell_ontology_class&#39;</span><span class="p">)[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">index</span>
<span class="p">)</span>

<span class="c1"># Personalizar etiquetas</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Count distribution per cell type and subtissue&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Cell type&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Quanrity&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Tissue&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/ea7147a2e104d7d3d75e310b6458015641c3e4c020c8ff1bb18bf962ca298211.png" src="_images/ea7147a2e104d7d3d75e310b6458015641c3e4c020c8ff1bb18bf962ca298211.png" />
</div>
</div>
</section>
<section id="heatmap-of-gene-expression">
<h2>Heatmap of gene expression<a class="headerlink" href="#heatmap-of-gene-expression" title="Link to this heading">#</a></h2>
<p>This heatmap represents the gene expression levels of a randomly selected subset of 20 genes across 20 different cells. The color intensity in the heatmap indicates the level of gene expression, where:</p>
<ul class="simple">
<li><p>Dark blue (low values) represents little to no gene expression.</p></li>
<li><p>Light blue to red (higher values) indicates increased expression levels, with red showing the highest expression values.</p></li>
</ul>
<p>There are 4 genes that have a representative high level of expresion into some of the cells, being HNRNPH2 with the hihgest level of expression for an specific one and repetitively high expression grades for more than one copared to the others.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Random Gene and cell selection</span>
<span class="n">sample_genes</span> <span class="o">=</span> <span class="n">df_counts</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># We select randomly 20 genes</span>
<span class="n">sample_cells</span> <span class="o">=</span> <span class="n">sample_genes</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>  <span class="c1"># We select randomly 20 cells</span>

<span class="c1"># Heat map</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">sample_cells</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="n">linewidths</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Heatmap of Gene Expression (Sample of 20 Genes and 20 Cells)&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Genes&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Cells&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/7de3e1f93b60e6fbfe7ad05af789b029b22393d92b9f6fe2390e8ded3082112a.png" src="_images/7de3e1f93b60e6fbfe7ad05af789b029b22393d92b9f6fe2390e8ded3082112a.png" />
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="glossary.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Important Concepts and Metrics Used for scRNA-seq Analysis</p>
      </div>
    </a>
    <a class="right-next"
       href="EDA_sparse_matrix.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Exploratory Data Analysis for the sparse matrix</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Exploratory Data Analysis for Metadata</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#import-data">Import data</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#analyze-and-describe-data">Analyze and Describe data</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-for-cell-type-and-proportion">Count for cell type and proportion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-and-proportion-of-subtissue">Count and proportion of subtissue</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#missing-data">Missing data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#statistical-summary">Statistical summary</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analysis-for-each-sex">Analysis For each sex</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-and-proportion-of-tissues-type-by-sex">Count and proportion of tissues type by sex</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cell-count-by-type">Cell count by type</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cell-count-by-tissue-type">Cell count by tissue type</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#heatmap-of-gene-expression">Heatmap of gene expression</a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Henry D. Saenz
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>