
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Important Concepts and Metrics Used for scRNA-seq Analysis &#8212; scRNA Sequencing Analysis</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'glossary';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Exploratory Data Analysis for Metadata" href="EDA_metadata.html" />
    <link rel="prev" title="Introduction to Single-Cell RNA Sequencing" href="introduction.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo_uninorte.png" class="logo__image only-light" alt="scRNA Sequencing Analysis - Home"/>
    <script>document.write(`<img src="_static/logo_uninorte.png" class="logo__image only-dark" alt="scRNA Sequencing Analysis - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Final Project - Deep Learning
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction to Single-Cell RNA Sequencing</a></li>

<li class="toctree-l1 current active"><a class="current reference internal" href="#">Important Concepts and Metrics Used for scRNA-seq Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="EDA_metadata.html">Exploratory Data Analysis for Metadata</a></li>

<li class="toctree-l1"><a class="reference internal" href="EDA_sparse_matrix.html">Exploratory Data Analysis for the sparse matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="Datos_celulares_y_geneticos.html">Exploratory Data Analysis for Glioblastoma data</a></li>
<li class="toctree-l1"><a class="reference internal" href="Agrupaciones_y_clusters.html">Exploratory Data Analysis for clusterization and gene expressions</a></li>
<li class="toctree-l1"><a class="reference internal" href="EDA_test.html">Exploratory Data Analysis for test data</a></li>

<li class="toctree-l1"><a class="reference internal" href="preprocessing.html">Preprocessing of Raw Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="PCA.html">Principal Component Analysis - Dimensionality reduction</a></li>

<li class="toctree-l1"><a class="reference internal" href="Benchmark_Models.html">Benchmark Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="spaco.html">Spatial Component Analysis Based on Latent Representations (SpaCO)</a></li>

</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/henrysaenz/eda_rna" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/henrysaenz/eda_rna/issues/new?title=Issue%20on%20page%20%2Fglossary.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/glossary.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Important Concepts and Metrics Used for scRNA-seq Analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#concepts-in-the-scrna-seq-analysis">Concepts in the scRNA - seq Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#key-metrics-in-scrna-seq-analysis">Key Metrics in scRNA seq Analysis</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="important-concepts-and-metrics-used-for-scrna-seq-analysis">
<h1>Important Concepts and Metrics Used for scRNA-seq Analysis<a class="headerlink" href="#important-concepts-and-metrics-used-for-scrna-seq-analysis" title="Link to this heading">#</a></h1>
<section id="concepts-in-the-scrna-seq-analysis">
<h2>Concepts in the scRNA - seq Analysis<a class="headerlink" href="#concepts-in-the-scrna-seq-analysis" title="Link to this heading">#</a></h2>
<div class="admonition-gene admonition">
<p class="admonition-title">Gene</p>
<p>A <strong>gene</strong> is a sequence of DNA that contains the instructions to produce a specific molecule, typically a protein or functional RNA. In scRNA-seq, gene expression is quantified by counting the RNA molecules transcribed from genes.</p>
</div>
<div class="admonition-cells admonition">
<p class="admonition-title">Cells</p>
<p><strong>Cells</strong> are the basic structural and functional units of all living organisms. In scRNA-seq, each cell is analyzed individually to understand its gene expression profile.</p>
</div>
<div class="admonition-rna admonition">
<p class="admonition-title">RNA</p>
<p><strong>RNA (Ribonucleic Acid)</strong> is a nucleic acid that plays essential roles in coding, decoding, regulation, and expression of genes. scRNA-seq measures the amount of RNA in each cell to infer gene expression.</p>
</div>
<div class="admonition-apoptosis admonition">
<p class="admonition-title">Apoptosis</p>
<p><strong>Apoptosis</strong> is a form of programmed cell death that occurs in multicellular organisms. It is characterized by specific cellular changes such as cell shrinkage, chromatin condensation, and membrane blebbing.</p>
</div>
<div class="admonition-apoptotic-cell admonition">
<p class="admonition-title">Apoptotic Cell</p>
<p>An <strong>apoptotic cell</strong> is a cell undergoing apoptosis. In scRNA-seq, high mitochondrial gene expression is often used as an indicator of apoptosis.</p>
</div>
<div class="admonition-gene-detection-rate admonition">
<p class="admonition-title">Gene Detection Rate</p>
<p>The <strong>Gene Detection Rate</strong> is the proportion of genes detected (i.e., with non-zero expression) in a given cell. It serves as a quality metric.</p>
<p><strong>Formula:</strong>
$<span class="math notranslate nohighlight">\(
\text{Gene Detection Rate} = \frac{\text{Number of detected genes in a cell}}{\text{Total number of genes}}
\)</span>$</p>
</div>
<div class="admonition-mitochondrial-cells admonition">
<p class="admonition-title">Mitochondrial Cells</p>
<p>In the context of scRNA-seq, <strong>mitochondrial cells</strong> usually refer to cells with abnormally high proportions of reads mapped to mitochondrial genes, often considered stressed or dying.</p>
</div>
<div class="admonition-mitochondrial-proportion admonition">
<p class="admonition-title">Mitochondrial Proportion</p>
<p>The <strong>Mitochondrial Proportion</strong> is the percentage of total reads in a cell that map to mitochondrial genes. High values suggest cell stress or apoptosis.</p>
<p><strong>Formula:</strong>
$<span class="math notranslate nohighlight">\(
\text{Mitochondrial Proportion} = \frac{\text{Mitochondrial gene counts}}{\text{Total gene counts}} \times 100
\)</span>$</p>
</div>
<div class="admonition-umi-counts admonition">
<p class="admonition-title">UMI Counts</p>
<p><strong>Unique Molecular Identifiers (UMIs)</strong> are short sequences used to count unique RNA molecules, minimizing amplification bias during PCR. Total UMI counts per cell indicate its transcriptome size.</p>
</div>
<div class="admonition-batch admonition">
<p class="admonition-title">Batch</p>
<p>A <strong>batch</strong> refers to a group of cells that have been processed together under the same experimental conditions. Variability between batches is common and can introduce confounding effects.</p>
</div>
<div class="admonition-batch-effect admonition">
<p class="admonition-title">Batch Effect</p>
<p><strong>Batch Effect</strong> is the unwanted variation in data due to technical differences between batches (e.g., reagent lots, sequencing runs), rather than biological differences.</p>
</div>
<div class="admonition-normalization admonition">
<p class="admonition-title">Normalization</p>
<p><strong>Normalization</strong> is the process of adjusting raw gene expression counts to account for differences in sequencing depth and other technical factors across cells.</p>
</div>
<div class="admonition-log-normalization admonition">
<p class="admonition-title">Log Normalization</p>
<p><strong>Log Normalization</strong> transforms the normalized counts using a logarithmic function to stabilize variance across genes.</p>
<p><strong>Formula:</strong>
$<span class="math notranslate nohighlight">\(
\text{LogNormalized}(x) = \log_2\left(\frac{x}{\text{total counts per cell}} \times \text{scale factor} + 1\right)
\)</span>$
Typically, the scale factor is 10,000.</p>
</div>
<div class="admonition-highly-variable-genes-hvgs admonition">
<p class="admonition-title">Highly Variable Genes (HVGs)</p>
<p><strong>Highly Variable Genes (HVGs)</strong> are genes that show greater expression variability across cells than expected by chance. HVGs are useful for downstream dimensionality reduction and clustering.</p>
</div>
<div class="admonition-principal-component-analysis-pca admonition">
<p class="admonition-title">Principal Component Analysis (PCA)</p>
<p><strong>PCA</strong> is a linear dimensionality reduction technique that transforms the data into a set of orthogonal axes (principal components), each capturing a portion of the variance.</p>
<p><strong>Formula:</strong>
PCA is computed by solving:
$<span class="math notranslate nohighlight">\(
\text{Cov}(X) = V \Lambda V^T
\)</span>$
Where:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(X\)</span>: mean-centered gene expression matrix</p></li>
<li><p><span class="math notranslate nohighlight">\(V\)</span>: eigenvectors (principal components)</p></li>
<li><p><span class="math notranslate nohighlight">\(\Lambda\)</span>: eigenvalues (explained variances)</p></li>
</ul>
</div>
</section>
<section id="key-metrics-in-scrna-seq-analysis">
<h2>Key Metrics in scRNA seq Analysis<a class="headerlink" href="#key-metrics-in-scrna-seq-analysis" title="Link to this heading">#</a></h2>
<div class="admonition-total-loss-zinb-loss admonition">
<p class="admonition-title">Total Loss (ZINB Loss)</p>
<p>Combines NB loss, dropout loss, and KL divergence with a weighting term ( \beta ).</p>
<p><strong>Formula:</strong></p>
<p>$<span class="math notranslate nohighlight">\(
\mathcal{L}_{Total} = \mathcal{L}_{NB} + \mathcal{L}_{Bern} + \beta \cdot \mathcal{L}_{KL}
\)</span>$```</p>
</div>
<div class="admonition-kl-divergence-kullback-leibler admonition">
<p class="admonition-title">KL Divergence (Kullback-Leibler)</p>
<p>Measures divergence between posterior <span class="math notranslate nohighlight">\(q(z|x)\)</span> and prior <span class="math notranslate nohighlight">\(p(z)\)</span>, used in VAEs.</p>
<p><strong>Formula:</strong></p>
<div class="math notranslate nohighlight">
\[
\mathcal{L}_{KL} = -\frac{1}{2} \sum_{i=1}^n \left( 1 + \log \sigma_i^2 - \mu_i^2 - \sigma_i^2 \right)
\]</div>
</div>
<div class="admonition-bernoulli-dropout-loss admonition">
<p class="admonition-title">Bernoulli Dropout Loss</p>
<p>Models dropout events as a Bernoulli process with probability <span class="math notranslate nohighlight">\(p_0\)</span>.</p>
<p><strong>Formula:</strong></p>
<div class="math notranslate nohighlight">
\[
\mathcal{L}_{Bern} = - \sum_{i,j} \left[ x_{ij} \log(1 - p_{0ij}) + (1 - x_{ij}) \log(p_{0ij}) \right]
\]</div>
</div>
<div class="admonition-negative-binomial-loss-nb admonition">
<p class="admonition-title">Negative Binomial Loss (NB)</p>
<p>Used to model count data with overdispersion. Assumes each gene follows a NB distribution.</p>
<p><strong>Formula:</strong></p>
<div class="math notranslate nohighlight">
\[
\mathcal{L}_{NB} = - \sum_{i,j} \left[ \log \Gamma(x_{ij} + \theta) - \log \Gamma(x_{ij} + 1) - \log \Gamma(\theta) + x_{ij} \log(\mu_{ij}) + \theta \log \left( \frac{\theta}{\theta + \mu_{ij}} \right) \right]
\]</div>
</div>
<div class="admonition-silhouette-score admonition">
<p class="admonition-title">Silhouette Score</p>
<p>Assesses how similar a cell is to its own cluster versus other clusters.</p>
<p><strong>Formula:</strong></p>
<div class="math notranslate nohighlight">
\[
s(i) = \frac{b(i) - a(i)}{max(a(i), b(i))}
\]</div>
<p>Where:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(a(i)\)</span> = mean intra-cluster distance</p></li>
<li><p><span class="math notranslate nohighlight">\(b(i)\)</span> = mean nearest-cluster distance</p></li>
</ul>
</div>
<div class="admonition-adjusted-rand-index-ari admonition">
<p class="admonition-title">Adjusted Rand Index (ARI)</p>
<p>Evaluates clustering similarity between predicted and true labels, adjusted for chance.</p>
<p><strong>Formula:</strong></p>
<div class="math notranslate nohighlight">
\[
ARI = \frac{RI - E[RI]}{max(RI) - E[RI]}
\]</div>
<p>Where <span class="math notranslate nohighlight">\(RI\)</span> is the Rand Index.</p>
</div>
<div class="admonition-normalized-mutual-information-nmi admonition">
<p class="admonition-title">Normalized Mutual Information (NMI)</p>
<p>Measures mutual dependence between predicted and true clusters.</p>
<p><strong>Formula:</strong></p>
<div class="math notranslate nohighlight">
\[
NMI = \frac{2 \cdot I(Y; \hat{Y})}{H(Y) + H(\hat{Y})}
\]</div>
<p>Where:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(I\)</span> is mutual information.</p></li>
<li><p><span class="math notranslate nohighlight">\(H\)</span> is entropy.</p></li>
</ul>
</div>
<div class="admonition-calinski-harabasz-index admonition">
<p class="admonition-title">Calinski-Harabasz Index</p>
<p>Cluster dispersion score; higher values indicate better-defined clusters.</p>
<p><strong>Formula:</strong></p>
<div class="math notranslate nohighlight">
\[
CH = \frac{Tr(B_k)}{k - 1} \cdot \frac{n - k}{Tr(W_k)}
\]</div>
<p>Where:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(Tr(B_k)\)</span> = between-cluster dispersion</p></li>
<li><p><span class="math notranslate nohighlight">\(Tr(W_k)\)</span> = within-cluster dispersion</p></li>
<li><p><span class="math notranslate nohighlight">\(k\)</span> = number of clusters</p></li>
<li><p><span class="math notranslate nohighlight">\(n\)</span> = total number of samples</p></li>
</ul>
</div>
<div class="admonition-pearson-correlation-coefficient admonition">
<p class="admonition-title">Pearson Correlation Coefficient</p>
<p>Measures linear correlation between real and reconstructed gene expressions.</p>
<p><strong>Formula:</strong></p>
<div class="math notranslate nohighlight">
\[
r = \frac{\sum_i (x_i - \bar{x})(\hat{x}_i - \bar{\hat{x}})}{\sqrt{\sum_i (x_i - \bar{x})^2 \sum_i (\hat{x}_i - \bar{\hat{x}})^2}}
\]</div>
</div>
<div class="admonition-reconstruction-loss-mse admonition">
<p class="admonition-title">Reconstruction Loss (MSE)</p>
<p>Basic loss function to measure average squared difference between original and reconstructed values.</p>
<p><strong>Formula:</strong></p>
<div class="math notranslate nohighlight">
\[
MSE = \frac{1}{n} \sum_{i=1}^{n} (x_i - \hat{x}_i)^2
\]</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="introduction.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Introduction to Single-Cell RNA Sequencing</p>
      </div>
    </a>
    <a class="right-next"
       href="EDA_metadata.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Exploratory Data Analysis for Metadata</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#concepts-in-the-scrna-seq-analysis">Concepts in the scRNA - seq Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#key-metrics-in-scrna-seq-analysis">Key Metrics in scRNA seq Analysis</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Henry D. Saenz
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      Â© Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>