
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Spatial Component Analysis Based on Latent Representations (SpaCO) &#8212; My sample book</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'spaco';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Benchmark Models" href="Benchmark_Models.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo_uninorte.png" class="logo__image only-light" alt="My sample book - Home"/>
    <script>document.write(`<img src="_static/logo_uninorte.png" class="logo__image only-dark" alt="My sample book - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Final Project - Deep Learning
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction to Single-Cell RNA Sequencing</a></li>

<li class="toctree-l1"><a class="reference internal" href="glossary.html">Important Concepts and Metrics Used for scRNA-seq Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="EDA_metadata.html">Exploratory Data Analysis for Metadata</a></li>

<li class="toctree-l1"><a class="reference internal" href="EDA_sparse_matrix.html">Exploratory Data Analysis for the sparse matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="Datos_celulares_y_geneticos.html">Exploratory Data Analysis for Glioblastoma data</a></li>
<li class="toctree-l1"><a class="reference internal" href="Agrupaciones_y_clusters.html">Exploratory Data Analysis for clusterization and gene expressions</a></li>
<li class="toctree-l1"><a class="reference internal" href="EDA_test.html">Exploratory Data Analysis for test data</a></li>

<li class="toctree-l1"><a class="reference internal" href="preprocessing.html">Preprocessing of Raw Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="PCA.html">Principal Component Analysis - Dimensionality reduction</a></li>

<li class="toctree-l1"><a class="reference internal" href="Benchmark_Models.html">Benchmark Models</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Spatial Component Analysis Based on Latent Representations (SpaCO)</a></li>

</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/henrysaenz/eda_rna" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/henrysaenz/eda_rna/issues/new?title=Issue%20on%20page%20%2Fspaco.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/spaco.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Spatial Component Analysis Based on Latent Representations (SpaCO)</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Spatial Component Analysis Based on Latent Representations (SpaCO)</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#variances">Variances</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#conclusion">Conclusion</a></li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="spatial-component-analysis-based-on-latent-representations-spaco">
<h1>Spatial Component Analysis Based on Latent Representations (SpaCO)<a class="headerlink" href="#spatial-component-analysis-based-on-latent-representations-spaco" title="Link to this heading">#</a></h1>
<p>To investigate the spatial organization of latent features learned by the VAE-GAT model, a Spectral Component Analysis (SpaCO) approach was applied over the latent space. This method allows identification of directions in the latent space that align with the topology of a cell‚Äìcell similarity graph.</p>
<p>Step-by-step process:</p>
<ol class="arabic simple">
<li><p>Latent Feature Extraction: The matrix <span class="math notranslate nohighlight">\(Œº ‚àà ‚Ñù‚ÅøÀ£·µà\)</span> was obtained by extracting the latent representations from the trained VAE-GAT model, where n is the number of cells and d is the latent dimensionality. This matrix Œº captures nonlinear features learned from gene expression data and the graph structure.</p></li>
<li><p>Graph Construction and Laplacian Normalization: A k-nearest neighbors (k-NN) graph was built over the cells using their expression profiles. From this, an adjacency matrix <span class="math notranslate nohighlight">\(W ‚àà ‚Ñù‚ÅøÀ£‚Åø\)</span> was constructed where <span class="math notranslate nohighlight">\(W[i,j] = 1\)</span> if cell j is a neighbor of cell i.
To model spatial dependencies, the symmetric normalized graph Laplacian L was computed using the formula:
$<span class="math notranslate nohighlight">\(L = D^{\frac{-1}{2}}WD^{\frac{-1}{2}}\)</span>$
where D is the degree matrix of the graph (diagonal matrix of node degrees). This normalization ensures scale invariance and improves the spectral analysis.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#in this case mu: latnet tensor of the model (shape: n_cells √ó latent_dim)</span>
<span class="c1"># (2 √ó num_edges)</span>
<span class="n">mu_np</span> <span class="o">=</span> <span class="n">mu</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
<span class="n">n_cells</span> <span class="o">=</span> <span class="n">mu_np</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># aDjacency matrix</span>
<span class="n">rows</span><span class="p">,</span> <span class="n">cols</span> <span class="o">=</span> <span class="n">edge_index_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">numpy</span><span class="p">(),</span> <span class="n">edge_index_cells</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rows</span><span class="p">))</span>
<span class="n">W</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">coo_matrix</span><span class="p">((</span><span class="n">data</span><span class="p">,</span> <span class="p">(</span><span class="n">rows</span><span class="p">,</span> <span class="n">cols</span><span class="p">)),</span> <span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="n">n_cells</span><span class="p">,</span> <span class="n">n_cells</span><span class="p">))</span><span class="o">.</span><span class="n">tocsc</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Symetric normalized Laplacian L matrix</span>
<span class="n">D</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">W</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
<span class="n">D_inv_sqrt</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="mf">1.0</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">D</span> <span class="o">+</span> <span class="mf">1e-8</span><span class="p">))</span>
<span class="n">L</span> <span class="o">=</span> <span class="n">D_inv_sqrt</span> <span class="o">@</span> <span class="n">W</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span> <span class="o">@</span> <span class="n">D_inv_sqrt</span>
</pre></div>
</div>
</div>
</div>
<ol class="arabic simple" start="3">
<li><p>Computation of the Spatial Covariance Matrix <span class="math notranslate nohighlight">\((Œº·µóLŒº)\)</span>: The spatially-weighted covariance matrix was computed as:
$<span class="math notranslate nohighlight">\(ùëÜ=ùúá^{ùëá}ùêøùúá\)</span>$</p></li>
</ol>
<p>This matrix reflects the alignment of latent features with the graph structure. Each entry <span class="math notranslate nohighlight">\(S[i,j]\)</span> measures how spatially correlated the ith and <span class="math notranslate nohighlight">\(jth\)</span> latent features are.</p>
<ol class="arabic simple" start="4">
<li><p>Eigen-Decomposition of the Spatial Covariance Matrix: The matrix S was diagonalized through eigen-decomposition:
$<span class="math notranslate nohighlight">\(ùëÜ=ùëâŒõùëâ^{ùëá}\)</span>$</p></li>
</ol>
<p>where Œõ is a diagonal matrix of eigenvalues, and V contains the corresponding eigenvectors. In linear algebra, eigenvectors represent directions (or axes) in the feature space along which the data exhibit a specific mode of variation; eigenvalues quantify the amount of variance captured along each eigenvector.</p>
<p>These eigenvectors are referred to as SpaCO components (SpaCs) ‚Äî analogous to principal components in PCA, but adapted to emphasize spatial coherence with the graph structure.</p>
<ol class="arabic simple" start="5">
<li><p>Projection onto SpaCO Components: The original latent matrix Œº was projected onto the top-k SpaCO components for <span class="math notranslate nohighlight">\(k ‚àà {1, ..., 30}\)</span> using:
$<span class="math notranslate nohighlight">\(ùëç_ùëò=ùúáùëâ_ùëò\)</span>$</p></li>
</ol>
<p>where <span class="math notranslate nohighlight">\(V_k\)</span> contains the first k eigenvectors. This yields new feature matrices <span class="math notranslate nohighlight">\(Z_k ‚àà ‚Ñù‚ÅøÀ£·µè\)</span> representing cells in the reduced SpaC space.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#  Œº·µóLŒº and eigenvectors</span>
<span class="n">spaco_matrix</span> <span class="o">=</span> <span class="n">mu_np</span><span class="o">.</span><span class="n">T</span> <span class="o">@</span> <span class="n">L</span> <span class="o">@</span> <span class="n">mu_np</span>
<span class="n">eigvals</span><span class="p">,</span> <span class="n">eigvecs</span> <span class="o">=</span> <span class="n">eigh</span><span class="p">(</span><span class="n">spaco_matrix</span><span class="p">)</span>
<span class="n">idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">eigvals</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="n">eigvals</span> <span class="o">=</span> <span class="n">eigvals</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
<span class="n">eigvecs</span> <span class="o">=</span> <span class="n">eigvecs</span><span class="p">[:,</span> <span class="n">idx</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>Clustering and Visualization: For each value of k, the following was performed:</p>
<ul class="simple">
<li><p>Neighborhood graph construction and UMAP embedding.</p></li>
<li><p>Clustering via the Leiden algorithm (resolution = 0.5).</p></li>
<li><p>Quantitative evaluation using the Silhouette Score (measuring intra/inter-cluster separation) and the Calinski‚ÄìHarabasz Index (measuring compactness and separation).</p></li>
<li><p>UMAP plots were generated to visualize the clustering results in 2D space.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># UMAP </span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">16</span><span class="p">))</span>
<span class="n">axs</span> <span class="o">=</span> <span class="n">axs</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

<span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">31</span><span class="p">):</span>
    <span class="n">spaco_features</span> <span class="o">=</span> <span class="n">mu_np</span> <span class="o">@</span> <span class="n">eigvecs</span><span class="p">[:,</span> <span class="p">:</span><span class="n">k</span><span class="p">]</span>  <span class="c1"># Projection over SpaCs</span>
    <span class="n">adata_spaco</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">spaco_features</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_spaco</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_spaco</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">adata_spaco</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

    <span class="n">labels</span> <span class="o">=</span> <span class="n">adata_spaco</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">umap_coords</span> <span class="o">=</span> <span class="n">adata_spaco</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap&quot;</span><span class="p">]</span>

    <span class="n">n_clusters</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">labels</span><span class="p">))</span>
    <span class="n">silhouette</span> <span class="o">=</span> <span class="n">silhouette_score</span><span class="p">(</span><span class="n">spaco_features</span><span class="p">,</span> <span class="n">labels</span><span class="p">)</span> <span class="k">if</span> <span class="n">n_clusters</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="k">else</span> <span class="o">-</span><span class="mi">1</span>
    <span class="n">calinski</span> <span class="o">=</span> <span class="n">calinski_harabasz_score</span><span class="p">(</span><span class="n">spaco_features</span><span class="p">,</span> <span class="n">labels</span><span class="p">)</span> <span class="k">if</span> <span class="n">n_clusters</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="k">else</span> <span class="o">-</span><span class="mi">1</span>

    <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
        <span class="s2">&quot;k&quot;</span><span class="p">:</span> <span class="n">k</span><span class="p">,</span>
        <span class="s2">&quot;n_clusters&quot;</span><span class="p">:</span> <span class="n">n_clusters</span><span class="p">,</span>
        <span class="s2">&quot;silhouette&quot;</span><span class="p">:</span> <span class="n">silhouette</span><span class="p">,</span>
        <span class="s2">&quot;calinski&quot;</span><span class="p">:</span> <span class="n">calinski</span>
    <span class="p">})</span>

    <span class="n">axs</span><span class="p">[</span><span class="n">k</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">umap_coords</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">umap_coords</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">c</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;tab20&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="n">k</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;SpaCs: </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="se">\n</span><span class="s2">Silh=</span><span class="si">{</span><span class="n">silhouette</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">, Cl=</span><span class="si">{</span><span class="n">n_clusters</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="n">k</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/90c97fe63d1f35ae11bc89641f711a0b7c44267bbb7012418dcac3b2deeb03a7.png" src="_images/90c97fe63d1f35ae11bc89641f711a0b7c44267bbb7012418dcac3b2deeb03a7.png" />
</div>
</div>
<p>Although projecting the latent space onto a single SpaCO component (k = 1) yielded the highest clustering metrics‚Äîachieving a Silhouette Score of 0.45 and a Calinski‚ÄìHarabasz Index exceeding 12,900‚Äîthis configuration produced 30 clusters, suggesting overfragmentation and limited interpretability. While the performance metrics were numerically optimal, relying on a single spatial component may oversimplify the underlying biological variability and overlook other spatially informative dimensions. In contrast, the projection onto the top five SpaCO components (k = 5) resulted in a more balanced outcome: 11 clusters with a respectable Silhouette Score of 0.316 and a Calinski Index of 1,350. This configuration captures richer spatial relationships while maintaining a manageable cluster granularity, making it more suitable for downstream biological interpretation. Therefore, although k = 1 defines an upper bound for clustering performance, k = 5 is considered the most practical and biologically meaningful choice in this context.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">results_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
<span class="n">results_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>k</th>
      <th>n_clusters</th>
      <th>silhouette</th>
      <th>calinski</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>30</td>
      <td>0.450049</td>
      <td>12927.626244</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>17</td>
      <td>0.312872</td>
      <td>3257.066478</td>
    </tr>
    <tr>
      <th>2</th>
      <td>3</td>
      <td>12</td>
      <td>0.310966</td>
      <td>1995.034782</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4</td>
      <td>11</td>
      <td>0.293294</td>
      <td>1584.206177</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5</td>
      <td>11</td>
      <td>0.316865</td>
      <td>1350.552981</td>
    </tr>
    <tr>
      <th>5</th>
      <td>6</td>
      <td>12</td>
      <td>0.292928</td>
      <td>1104.428820</td>
    </tr>
    <tr>
      <th>6</th>
      <td>7</td>
      <td>10</td>
      <td>0.276201</td>
      <td>1163.911760</td>
    </tr>
    <tr>
      <th>7</th>
      <td>8</td>
      <td>11</td>
      <td>0.280297</td>
      <td>1018.565714</td>
    </tr>
    <tr>
      <th>8</th>
      <td>9</td>
      <td>11</td>
      <td>0.275027</td>
      <td>947.401387</td>
    </tr>
    <tr>
      <th>9</th>
      <td>10</td>
      <td>12</td>
      <td>0.261968</td>
      <td>811.374206</td>
    </tr>
    <tr>
      <th>10</th>
      <td>11</td>
      <td>12</td>
      <td>0.256656</td>
      <td>771.557441</td>
    </tr>
    <tr>
      <th>11</th>
      <td>12</td>
      <td>12</td>
      <td>0.252305</td>
      <td>746.628839</td>
    </tr>
    <tr>
      <th>12</th>
      <td>13</td>
      <td>12</td>
      <td>0.245824</td>
      <td>717.766980</td>
    </tr>
    <tr>
      <th>13</th>
      <td>14</td>
      <td>12</td>
      <td>0.238587</td>
      <td>693.196194</td>
    </tr>
    <tr>
      <th>14</th>
      <td>15</td>
      <td>12</td>
      <td>0.231578</td>
      <td>652.603961</td>
    </tr>
    <tr>
      <th>15</th>
      <td>16</td>
      <td>12</td>
      <td>0.228483</td>
      <td>651.225874</td>
    </tr>
    <tr>
      <th>16</th>
      <td>17</td>
      <td>12</td>
      <td>0.226320</td>
      <td>636.158786</td>
    </tr>
    <tr>
      <th>17</th>
      <td>18</td>
      <td>12</td>
      <td>0.224532</td>
      <td>621.539355</td>
    </tr>
    <tr>
      <th>18</th>
      <td>19</td>
      <td>12</td>
      <td>0.221696</td>
      <td>610.255870</td>
    </tr>
    <tr>
      <th>19</th>
      <td>20</td>
      <td>12</td>
      <td>0.218620</td>
      <td>590.883966</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<section id="variances">
<h2>Variances<a class="headerlink" href="#variances" title="Link to this heading">#</a></h2>
<p>To assess the extent to which the learned latent representations capture spatially coherent structures, a variance decomposition analysis was performed. This approach aimed to separate the total variability in the latent space into two components: global variance and spatially structured variance, the latter being aligned with the topology of the cell‚Äìcell graph. The global variance quantifies the total spread or variability in the latent space, independent of any neighborhood relationships. In contrast, the spatial variance captures the portion of this variability that is consistent with the local structure encoded in the cell‚Äìcell graph. This graph represents biological proximity or similarity between cells and is typically constructed based on gene expression or spatial coordinates. By computing the fraction of the total variance that is spatially structured, it becomes possible to evaluate how much of the latent signal is influenced by the local graph topology. This ratio serves as an indicator of spatial alignment in the latent embedding, revealing whether the model has learned features that reflect biologically meaningful organization across the graph. A high spatial fraction suggests that the latent space encodes localized patterns, while a lower fraction would imply more global or unstructured variability.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">spaco_matrix</span> <span class="o">=</span> <span class="n">mu_np</span><span class="o">.</span><span class="n">T</span> <span class="o">@</span> <span class="n">L</span> <span class="o">@</span> <span class="n">mu_np</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">var_global</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">trace</span><span class="p">(</span><span class="n">mu_np</span><span class="o">.</span><span class="n">T</span> <span class="o">@</span> <span class="n">mu_np</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Spatial (graph-weighted) variance</span>
<span class="n">var_spatial</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">trace</span><span class="p">(</span><span class="n">mu_np</span><span class="o">.</span><span class="n">T</span> <span class="o">@</span> <span class="n">L</span> <span class="o">@</span> <span class="n">mu_np</span><span class="p">)</span>

<span class="c1"># Proportion of variance explained by spatial structure</span>
<span class="n">spatial_fraction</span> <span class="o">=</span> <span class="n">var_spatial</span> <span class="o">/</span> <span class="n">var_global</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Global Variance</strong> refers to the sum of the variances across all dimensions of the latent space Œº. It indicates the overall diversity among the cells in that space.
The global variance, calculated as 969,838, reflects the total spread of the latent representations across all dimensions of the embedding space. This value is relatively high, indicating that the model has captured a broad range of variability among the input cells. Such a scale suggests that the latent space is rich in structure and effectively separates diverse cell states or types. In practical terms, a high global variance often implies that the encoder of the VAE+GAT model is preserving significant biological or structural heterogeneity</p>
<div class="admonition-spatially-structured-variance admonition">
<p class="admonition-title">Spatially-Structured Variance</p>
<p>Computed as: trace(Œº·µó L Œº)
This value reflects the extent to which the variability in the latent space is organized in a way that is consistent with the cell‚Äìcell graph structure.
In other words, it measures whether the differences in the embedding align with the graph topology (i.e., neighborhood relationships).</p>
</div>
<p>In this case: 184,416, indicating that a significant portion of the variation is aligned with the graph.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Global variance: </span><span class="si">{</span><span class="n">var_global</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Spatial (graph-aware) variance: </span><span class="si">{</span><span class="n">var_spatial</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Fraction of spatially-structured variance: </span><span class="si">{</span><span class="n">spatial_fraction</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Global variance: 969838.9375
Spatial (graph-aware) variance: 184416.0143
Fraction of spatially-structured variance: 0.1902
</pre></div>
</div>
</div>
</div>
<div class="admonition-spatially-structured-fraction-0-19 admonition">
<p class="admonition-title">Spatially-Structured Fraction (~0.19)</p>
<p>This means that approximately 19% of the variance in the latent space exhibits a topological structure‚Äîi.e., it is ‚Äúlocalized‚Äù on the graph.</p>
<p>In other words, about 19% of the latent signal is correlated with the connectivity of the cell‚Äìcell graph.</p>
</div>
<p>How to interpret the 0.19 value?</p>
<p>This is not a low value. In single-cell data‚Äîespecially when not derived from spatial technologies (such as Visium or MERFISH)‚Äîit is common for only a small fraction of the variance to be spatially organized. This suggests that the VAE-GAT embedding has captured structure that respects the graph of neighboring cells, at least partially.</p>
<ul class="simple">
<li><p>The cell‚Äìcell graph is contributing meaningful information to the model.</p></li>
<li><p>The embedding is not random with respect to the graph.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">eigvals</span><span class="p">,</span> <span class="n">eigvecs</span> <span class="o">=</span> <span class="n">eigh</span><span class="p">(</span><span class="n">mu_np</span><span class="o">.</span><span class="n">T</span> <span class="o">@</span> <span class="n">L</span> <span class="o">@</span> <span class="n">mu_np</span><span class="p">)</span>
<span class="n">eigvals</span> <span class="o">=</span> <span class="n">eigvals</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">eigvals</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span>  <span class="c1"># sort descending</span>

<span class="c1"># Normalize by global variance to get relative contribution</span>
<span class="n">eigvals_relative</span> <span class="o">=</span> <span class="n">eigvals</span> <span class="o">/</span> <span class="n">var_global</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">eigvals</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">eigvals_relative</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;SpaC Component&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Fraction of spatially-explained variance&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Variance Explained by SpaC Components&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/2d677f677d0b64f7605d8c4bb8209346a018b721bfd53050964c5f2ef33a5cf0.png" src="_images/2d677f677d0b64f7605d8c4bb8209346a018b721bfd53050964c5f2ef33a5cf0.png" />
</div>
</div>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="conclusion">
<h1>Conclusion<a class="headerlink" href="#conclusion" title="Link to this heading">#</a></h1>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="Benchmark_Models.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Benchmark Models</p>
      </div>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Spatial Component Analysis Based on Latent Representations (SpaCO)</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#variances">Variances</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#conclusion">Conclusion</a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The Jupyter Book Community
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      ¬© Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>