
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Exploratory Data Analysis for clusterization and gene expressions &#8212; My sample book</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Agrupaciones_y_clusters';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Exploratory Data Analysis for test data" href="EDA_test.html" />
    <link rel="prev" title="Exploratory Data Analysis for Glioblastoma data" href="Datos_celulares_y_geneticos.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo_uninorte.png" class="logo__image only-light" alt="My sample book - Home"/>
    <script>document.write(`<img src="_static/logo_uninorte.png" class="logo__image only-dark" alt="My sample book - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Final Project - Deep Learning
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction to Single-Cell RNA Sequencing</a></li>

<li class="toctree-l1"><a class="reference internal" href="EDA_metadata.html">Exploratory Data Analysis for Metadata</a></li>

<li class="toctree-l1"><a class="reference internal" href="EDA_sparse_matrix.html">Exploratory Data Analysis for the sparse matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="Datos_celulares_y_geneticos.html">Exploratory Data Analysis for Glioblastoma data</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Exploratory Data Analysis for clusterization and gene expressions</a></li>
<li class="toctree-l1"><a class="reference internal" href="EDA_test.html">Exploratory Data Analysis for test data</a></li>

</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/henrysaenz/eda_rna" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/henrysaenz/eda_rna/issues/new?title=Issue%20on%20page%20%2FAgrupaciones_y_clusters.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/Agrupaciones_y_clusters.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Exploratory Data Analysis for clusterization and gene expressions</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#import-data-and-information">Import Data and Information</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#adata-pbmc3k">adata_pbmc3k</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#variables-in-the-dataset">üîç Variables in the dataset</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">üîç Variables in the dataset</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#total-rna-counts-vs-mitochondrial-percentage">Total RNA Counts vs. Mitochondrial Percentage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#number-of-genes-detected-vs-mitochondrial-percentage">Number of Genes Detected vs. Mitochondrial Percentage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#violin-plot-of-total-gene-counts-by-clusters-louvain">Violin Plot of Total Gene Counts by Clusters (louvain)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-xpression-per-cluster">Gene xpression per cluster</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#correlation-matrix">Correlation matrix</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genes-vs-mean-expression">Genes vs. Mean Expression</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#percentage-dropout-vs-mean-counts">Percentage Dropout vs. Mean Counts</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#distribution-of-more-expressed-genes-per-cluster">Distribution of more expressed genes per cluster</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#expression-of-a-specific-gene-across-clusters">Expression of a Specific Gene Across Clusters</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clusterization">Clusterization</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interpretation">üîç Interpretation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#observations"><strong>Observations</strong></a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="exploratory-data-analysis-for-clusterization-and-gene-expressions">
<h1>Exploratory Data Analysis for clusterization and gene expressions<a class="headerlink" href="#exploratory-data-analysis-for-clusterization-and-gene-expressions" title="Link to this heading">#</a></h1>
<section id="import-data-and-information">
<h2>Import Data and Information<a class="headerlink" href="#import-data-and-information" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">IPython.core.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">display</span><span class="p">,</span> <span class="n">HTML</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_path</span> <span class="o">=</span> <span class="s2">&quot;C:/Users/henry/Desktop/python/jbook_rna/datos/Single-cell RNASeq data from Mouse Brain&quot;</span>
<span class="n">data_path2</span> <span class="o">=</span> <span class="s2">&quot;C:/Users/henry/Desktop/python/jbook_rna/datos/Theory - Intro to single-cell RNAseq Images&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">glioblastoma_norm_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="s2">&quot;glioblastoma_normalized.h5ad&quot;</span><span class="p">)</span>
<span class="n">glioblastoma_raw_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="s2">&quot;glioblastoma_raw.h5ad&quot;</span><span class="p">)</span>
<span class="n">pbmc3k_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="s2">&quot;pbmc3k.h5ad&quot;</span><span class="p">)</span>

<span class="n">adata_glioblastoma_norm</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">glioblastoma_norm_file</span><span class="p">)</span>
<span class="n">adata_glioblastoma_raw</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">glioblastoma_raw_file</span><span class="p">)</span>
<span class="n">adata_pbmc3k</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">pbmc3k_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="adata-pbmc3k">
<h2>adata_pbmc3k<a class="headerlink" href="#adata-pbmc3k" title="Link to this heading">#</a></h2>
<p>This dataset contains information about gene expression in peripheral blood mononuclear cells (PBMC). Each row represents a gene, and the columns contain statistics about its expression across different cells.</p>
</section>
<section id="variables-in-the-dataset">
<h2>üîç Variables in the dataset<a class="headerlink" href="#variables-in-the-dataset" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p><strong>Variable</strong></p></th>
<th class="head"><p><strong>Description</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">gene_ids</span></code></p></td>
<td><p>Unique gene identifier in Ensembl format (<code class="docutils literal notranslate"><span class="pre">ENSGxxxxxxxxxxx</span></code>).</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">n_cells_by_counts</span></code></p></td>
<td><p>Number of cells in which the gene was detected.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">mean_counts</span></code></p></td>
<td><p>Mean expression value of the gene across all cells.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">log1p_mean_counts</span></code></p></td>
<td><p>Log-transformed (<code class="docutils literal notranslate"><span class="pre">log1p</span></code>) mean expression of the gene.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">pct_dropout_by_counts</span></code></p></td>
<td><p>Percentage of cells where the gene was not expressed (<code class="docutils literal notranslate"><span class="pre">dropout</span></code>).</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">total_counts</span></code></p></td>
<td><p>Total sum of expression values for the gene across all cells.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">log1p_total_counts</span></code></p></td>
<td><p>Log-transformed (<code class="docutils literal notranslate"><span class="pre">log1p</span></code>) total expression count.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">n_cells</span></code></p></td>
<td><p>Total number of cells in which gene expression was measured.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">highly_variable</span></code></p></td>
<td><p>Indicates whether the gene is highly variable (<code class="docutils literal notranslate"><span class="pre">True/False</span></code>), useful for feature selection.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">means</span></code></p></td>
<td><p>Mean gene expression in the sample, without log transformation.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">dispersions</span></code></p></td>
<td><p>Dispersion of the gene expression across all cells.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">dispersions_norm</span></code></p></td>
<td><p>Normalized dispersion of the gene, adjusted by the mean.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_pbmc3k</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_ids</th>
      <th>n_cells_by_counts</th>
      <th>mean_counts</th>
      <th>log1p_mean_counts</th>
      <th>pct_dropout_by_counts</th>
      <th>total_counts</th>
      <th>log1p_total_counts</th>
      <th>n_cells</th>
      <th>highly_variable</th>
      <th>means</th>
      <th>dispersions</th>
      <th>dispersions_norm</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TNFRSF4</th>
      <td>ENSG00000186827</td>
      <td>155</td>
      <td>0.077407</td>
      <td>0.074558</td>
      <td>94.259259</td>
      <td>209.0</td>
      <td>5.347107</td>
      <td>155</td>
      <td>True</td>
      <td>0.277410</td>
      <td>2.086050</td>
      <td>0.665406</td>
    </tr>
    <tr>
      <th>CPSF3L</th>
      <td>ENSG00000127054</td>
      <td>202</td>
      <td>0.094815</td>
      <td>0.090585</td>
      <td>92.518519</td>
      <td>256.0</td>
      <td>5.549076</td>
      <td>202</td>
      <td>True</td>
      <td>0.385194</td>
      <td>4.506988</td>
      <td>2.955005</td>
    </tr>
    <tr>
      <th>ATAD3C</th>
      <td>ENSG00000215915</td>
      <td>9</td>
      <td>0.009259</td>
      <td>0.009217</td>
      <td>99.666667</td>
      <td>25.0</td>
      <td>3.258096</td>
      <td>9</td>
      <td>True</td>
      <td>0.038252</td>
      <td>3.953486</td>
      <td>4.352607</td>
    </tr>
    <tr>
      <th>C1orf86</th>
      <td>ENSG00000162585</td>
      <td>501</td>
      <td>0.227778</td>
      <td>0.205206</td>
      <td>81.444444</td>
      <td>615.0</td>
      <td>6.423247</td>
      <td>501</td>
      <td>True</td>
      <td>0.678283</td>
      <td>2.713521</td>
      <td>0.543182</td>
    </tr>
    <tr>
      <th>RER1</th>
      <td>ENSG00000157916</td>
      <td>608</td>
      <td>0.298148</td>
      <td>0.260939</td>
      <td>77.481481</td>
      <td>805.0</td>
      <td>6.692084</td>
      <td>608</td>
      <td>True</td>
      <td>0.814813</td>
      <td>3.447533</td>
      <td>1.582528</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="id1">
<h2>üîç Variables in the dataset<a class="headerlink" href="#id1" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p><strong>Variable</strong></p></th>
<th class="head"><p><strong>Description</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">n_genes_by_counts</span></code></p></td>
<td><p>Number of genes detected in the cell.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">log1p_n_genes_by_counts</span></code></p></td>
<td><p>Log-transformed (<code class="docutils literal notranslate"><span class="pre">log1p</span></code>) number of detected genes.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">total_counts</span></code></p></td>
<td><p>Total number of RNA molecule counts detected in the cell.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">log1p_total_counts</span></code></p></td>
<td><p>Log-transformed (<code class="docutils literal notranslate"><span class="pre">log1p</span></code>) total RNA counts.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">pct_counts_in_top_50_genes</span></code></p></td>
<td><p>Percentage of total RNA counts contributed by the <strong>top 50 most expressed genes</strong> in the cell.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">pct_counts_in_top_100_genes</span></code></p></td>
<td><p>Percentage of total RNA counts contributed by the <strong>top 100 most expressed genes</strong> in the cell.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">pct_counts_in_top_200_genes</span></code></p></td>
<td><p>Percentage of total RNA counts contributed by the <strong>top 200 most expressed genes</strong> in the cell.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">pct_counts_in_top_500_genes</span></code></p></td>
<td><p>Percentage of total RNA counts contributed by the <strong>top 500 most expressed genes</strong> in the cell.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">n_genes</span></code></p></td>
<td><p>Number of genes detected in the cell (<strong>alternative representation of <code class="docutils literal notranslate"><span class="pre">n_genes_by_counts</span></code></strong>).</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">percent_mito</span></code></p></td>
<td><p>Percentage of mitochondrial gene expression out of the total RNA counts (<strong>used to assess cell quality</strong>).</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">n_counts</span></code></p></td>
<td><p>Total number of UMI (Unique Molecular Identifiers) counts per cell (<strong>alternative representation of <code class="docutils literal notranslate"><span class="pre">total_counts</span></code></strong>).</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">louvain</span></code></p></td>
<td><p>Cluster assignment using the <strong>Louvain community detection algorithm</strong> (unsupervised clustering).</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">leiden</span></code></p></td>
<td><p>Cluster assignment using the <strong>Leiden clustering algorithm</strong>, another method for detecting groups of similar cells.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>n_genes_by_counts</th>
      <th>log1p_n_genes_by_counts</th>
      <th>total_counts</th>
      <th>log1p_total_counts</th>
      <th>pct_counts_in_top_50_genes</th>
      <th>pct_counts_in_top_100_genes</th>
      <th>pct_counts_in_top_200_genes</th>
      <th>pct_counts_in_top_500_genes</th>
      <th>n_genes</th>
      <th>percent_mito</th>
      <th>n_counts</th>
      <th>louvain</th>
      <th>leiden</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACATACAACCAC-1</th>
      <td>781</td>
      <td>6.661855</td>
      <td>2421.0</td>
      <td>7.792349</td>
      <td>47.748864</td>
      <td>63.279637</td>
      <td>74.969021</td>
      <td>88.393226</td>
      <td>781</td>
      <td>0.030178</td>
      <td>2419.0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>AAACATTGAGCTAC-1</th>
      <td>1352</td>
      <td>7.210080</td>
      <td>4903.0</td>
      <td>8.497807</td>
      <td>45.502753</td>
      <td>61.023863</td>
      <td>71.813176</td>
      <td>82.622884</td>
      <td>1352</td>
      <td>0.037936</td>
      <td>4903.0</td>
      <td>2</td>
      <td>2</td>
    </tr>
    <tr>
      <th>AAACATTGATCAGC-1</th>
      <td>1131</td>
      <td>7.031741</td>
      <td>3149.0</td>
      <td>8.055158</td>
      <td>41.314703</td>
      <td>53.794856</td>
      <td>65.449349</td>
      <td>79.961893</td>
      <td>1131</td>
      <td>0.008897</td>
      <td>3147.0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>AAACCGTGCTTCCG-1</th>
      <td>960</td>
      <td>6.867974</td>
      <td>2639.0</td>
      <td>7.878534</td>
      <td>39.029936</td>
      <td>52.898825</td>
      <td>66.691929</td>
      <td>82.569155</td>
      <td>960</td>
      <td>0.017431</td>
      <td>2639.0</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>AAACCGTGTATGCG-1</th>
      <td>522</td>
      <td>6.259581</td>
      <td>981.0</td>
      <td>6.889591</td>
      <td>44.852192</td>
      <td>55.657492</td>
      <td>67.176351</td>
      <td>97.757390</td>
      <td>522</td>
      <td>0.012245</td>
      <td>980.0</td>
      <td>4</td>
      <td>4</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="total-rna-counts-vs-mitochondrial-percentage">
<h2>Total RNA Counts vs. Mitochondrial Percentage<a class="headerlink" href="#total-rna-counts-vs-mitochondrial-percentage" title="Link to this heading">#</a></h2>
<p>High mitochondrial content may indicate stressed or dying cells.</p>
<ol class="arabic simple">
<li><p><strong>Dense Cluster Around Low Total Counts</strong></p>
<ul class="simple">
<li><p>The majority of data points are concentrated around <strong>total counts between 1000 and 4000</strong>.</p></li>
<li><p>The mitochondrial percentage in this range is <strong>mostly between 1% and 4%</strong>.</p></li>
</ul>
</li>
<li><p><strong>Increase in Variability at Higher Total Counts</strong></p>
<ul class="simple">
<li><p>As <code class="docutils literal notranslate"><span class="pre">total_counts</span></code> increases beyond <strong>4000</strong>, there is <strong>greater variability</strong> in mitochondrial percentage.</p></li>
<li><p>Some cells have <strong>higher mitochondrial percentages (~5%)</strong>, but most stay below this level.</p></li>
</ul>
</li>
<li><p><strong>No Strong Correlation</strong></p>
<ul class="simple">
<li><p>The points are widely scattered, indicating <strong>no clear linear relationship</strong> between total counts and mitochondrial percentage.</p></li>
<li><p>This suggests that mitochondrial gene expression is not strictly dependent on total RNA counts but may be influenced by other biological or technical factors.</p></li>
</ul>
</li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">adata_pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;total_counts&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;percent_mito&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Total Counts vs. Mitochondrial Percentage&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/94baa957c270cdad47c3df9379b93c83a2b29b12dbf9c546447cf6a1571425fe.png" src="_images/94baa957c270cdad47c3df9379b93c83a2b29b12dbf9c546447cf6a1571425fe.png" />
</div>
</div>
</section>
<section id="number-of-genes-detected-vs-mitochondrial-percentage">
<h2>Number of Genes Detected vs. Mitochondrial Percentage<a class="headerlink" href="#number-of-genes-detected-vs-mitochondrial-percentage" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p><strong>Dense Cluster Around 500-1000 Genes</strong></p>
<ul class="simple">
<li><p>Most cells have detected <strong>between 500 and 1000 genes</strong>.</p></li>
<li><p>The mitochondrial percentage in this range <strong>varies between 1% and 4%</strong>, with some outliers reaching <strong>above 5%</strong>.</p></li>
</ul>
</li>
<li><p><strong>Increase in Variability for Higher Gene Counts</strong></p>
<ul class="simple">
<li><p>As <code class="docutils literal notranslate"><span class="pre">n_genes_by_counts</span></code> increases beyond <strong>1500</strong>, the variability in mitochondrial percentage expands.</p></li>
<li><p>Some cells with <strong>&gt;2000 detected genes</strong> have <strong>very low mitochondrial percentages</strong>, while others have higher values.</p></li>
</ul>
</li>
<li><p><strong>No Strong Correlation</strong></p>
<ul class="simple">
<li><p>The scatter of points suggests <strong>no clear linear trend</strong> between the number of detected genes and mitochondrial gene expression.</p></li>
<li><p>A weak trend may indicate that cells with fewer detected genes tend to have slightly higher mitochondrial percentages, which could be an indicator of low-quality or stressed cells.</p></li>
</ul>
</li>
</ol>
<ul class="simple">
<li><p><strong>Cells with lower detected genes and higher mitochondrial content</strong> might represent <strong>stressed or dying cells</strong> with degraded cytoplasmic RNA.</p></li>
<li><p><strong>No strong correlation</strong> exists, though cells with very low gene detection may show slightly elevated mitochondrial expression.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">adata_pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;percent_mito&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Number of Genes Detected vs. Mitochondrial Percentage&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/077e60b97981c31eea17e77d3cc4b01db5ede8715ac6d989a9913a01c6089440.png" src="_images/077e60b97981c31eea17e77d3cc4b01db5ede8715ac6d989a9913a01c6089440.png" />
</div>
</div>
</section>
<section id="violin-plot-of-total-gene-counts-by-clusters-louvain">
<h2>Violin Plot of Total Gene Counts by Clusters (louvain)<a class="headerlink" href="#violin-plot-of-total-gene-counts-by-clusters-louvain" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p><strong>Clusters with higher total RNA counts (e.g., Cluster 6)</strong> may represent <strong>highly transcriptionally active cells</strong> or certain cell types with more RNA content.</p></li>
<li><p><strong>Clusters with lower RNA counts (e.g., Cluster 7)</strong> might include <strong>low-quality cells, stressed cells, or specific cell types with inherently low RNA expression</strong>.</p></li>
<li><p><strong>Outliers with extreme RNA counts</strong> should be examined further to determine whether they are biologically meaningful or artifacts of data collection.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">adata_pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;louvain&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;total_counts&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Distribution of Total RNA Counts per Cluster&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/29d0d1d1b8e9609fc6f3f36eaf38e84d9f94f49faf7c8fa889a545fe391b10c2.png" src="_images/29d0d1d1b8e9609fc6f3f36eaf38e84d9f94f49faf7c8fa889a545fe391b10c2.png" />
</div>
</div>
</section>
<section id="gene-xpression-per-cluster">
<h2>Gene xpression per cluster<a class="headerlink" href="#gene-xpression-per-cluster" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p>Clusters <strong>1, 4, 5, and 6 show greater dispersion</strong> in the number of expressed genes, indicating <strong>heterogeneity in transcriptional activity</strong> within these groups.</p></li>
<li><p>The presence of <strong>outliers</strong> (points outside the whiskers of the box) in all clusters suggests the existence of some cells with a number of expressed genes <strong>significantly higher or lower than the majority</strong>.</p></li>
<li><p>There are notable differences in the number of expressed genes between clusters, indicating <strong>cellular heterogeneity</strong>.</p></li>
<li><p><strong>Cluster 6 has the greatest diversity of gene expression</strong>, while <strong>Cluster 7 has the least</strong>.</p></li>
<li><p>It is advisable to conduct further analysis to determine if the <strong>clusters with a low number of expressed genes contain viable cells or if they represent experimental artifacts</strong>.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;louvain&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">adata_pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Number of genes expressed per cluster (Louvain)&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/71a9e4bf296bdee83aa62c8533f8a6cf6bcbe6e8cb278db3b373a44ea6da2afa.png" src="_images/71a9e4bf296bdee83aa62c8533f8a6cf6bcbe6e8cb278db3b373a44ea6da2afa.png" />
</div>
</div>
</section>
<section id="correlation-matrix">
<h2>Correlation matrix<a class="headerlink" href="#correlation-matrix" title="Link to this heading">#</a></h2>
<p><strong>Weak negative correlation between <code class="docutils literal notranslate"><span class="pre">percent_mito</span></code> and other variables (<code class="docutils literal notranslate"><span class="pre">r</span> <span class="pre">‚âà</span> <span class="pre">-0.04</span></code>)</strong>:</p>
<ul class="simple">
<li><p>Total number of genes or the total amount of RNA in the cell.</p></li>
<li><p>A value close to zero suggests that <strong>differences in mitochondrial expression are not strongly linked to the total amount of detected genetic material</strong>.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">corr_matrix</span> <span class="o">=</span> <span class="n">adata_pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;total_counts&quot;</span><span class="p">,</span> <span class="s2">&quot;percent_mito&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">corr</span><span class="p">()</span>
<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">corr_matrix</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Correlation Heatmap (obs)&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/87a028b8172adc4f04e7b4851a5d3ac96e610cb9d805a42c865dec063c6b3c5a.png" src="_images/87a028b8172adc4f04e7b4851a5d3ac96e610cb9d805a42c865dec063c6b3c5a.png" />
</div>
</div>
</section>
<section id="genes-vs-mean-expression">
<h2>Genes vs. Mean Expression<a class="headerlink" href="#genes-vs-mean-expression" title="Link to this heading">#</a></h2>
<p>The plot shows a negative correlation between expression level and variability, influenced by dropout effects, biological heterogeneity, and overdispersion.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">adata_pbmc3k</span><span class="o">.</span><span class="n">var</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;means&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;dispersions_norm&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Mean Expression&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Normalized Variability&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Relationship between Variability and Expression&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/33d65a442eb29dbb62845cb249726842bf5378b62d5c34aa8f0bf09e3d48628c.png" src="_images/33d65a442eb29dbb62845cb249726842bf5378b62d5c34aa8f0bf09e3d48628c.png" />
</div>
</div>
</section>
<section id="percentage-dropout-vs-mean-counts">
<h2>Percentage Dropout vs. Mean Counts<a class="headerlink" href="#percentage-dropout-vs-mean-counts" title="Link to this heading">#</a></h2>
<p>Higher dropout (zero expression in many cells) means a gene is detected in fewer cells. This plot highlights the dropout effect in scRNA-seq and suggests that lowly expressed genes should be interpreted cautiously due to potential detection bias.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">adata_pbmc3k</span><span class="o">.</span><span class="n">var</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;mean_counts&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;pct_dropout_by_counts&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Gene Dropout vs. Mean Expression&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/05f513045e6441ffd1f96a444e19df60205a35f3f26422ce63ea1e128a2a260a.png" src="_images/05f513045e6441ffd1f96a444e19df60205a35f3f26422ce63ea1e128a2a260a.png" />
</div>
</div>
</section>
<section id="distribution-of-more-expressed-genes-per-cluster">
<h2>Distribution of more expressed genes per cluster<a class="headerlink" href="#distribution-of-more-expressed-genes-per-cluster" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p>Cluster 7 exhibits a much wider spread, with some cells having over 70% of their total counts coming from the top 50 genes, indicating a stronger dominance of a few genes in these cells.</p></li>
<li><p>Most clusters have a percentage of counts concentrated around 40-50%, suggesting that a significant proportion of gene expression is dominated by a few highly expressed genes.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">adata_pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;louvain&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;pct_counts_in_top_50_genes&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Percentage of Counts in the Top 50 Most Expressed Genes&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/e4dc2f08ea507f07a6744d8f6ea09b5d4c1ef58181ee2dd88e477fac0e2c6c69.png" src="_images/e4dc2f08ea507f07a6744d8f6ea09b5d4c1ef58181ee2dd88e477fac0e2c6c69.png" />
</div>
</div>
</section>
<section id="expression-of-a-specific-gene-across-clusters">
<h2>Expression of a Specific Gene Across Clusters<a class="headerlink" href="#expression-of-a-specific-gene-across-clusters" title="Link to this heading">#</a></h2>
<p>Choose a highly variable gene, e.g., ‚ÄúTNFRSF4‚Äù, and plot its expression:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene</span> <span class="o">=</span> <span class="s2">&quot;TNFRSF4&quot;</span>
<span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">adata_pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;louvain&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">adata_pbmc3k</span><span class="p">[:,</span> <span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Expression of </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2"> Across Clusters&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/21a26f4d11ea27af0f5f7edc4f8f55f4c1951f21f89399208850b5230dd2f6a0.png" src="_images/21a26f4d11ea27af0f5f7edc4f8f55f4c1951f21f89399208850b5230dd2f6a0.png" />
</div>
</div>
<ul class="simple">
<li><p>Cluster 0 exhibits the widest range of expression values, with some cells showing high expression levels (outliers).</p></li>
<li><p>Most clusters (e.g., 1, 2, 3, 4, 5, 6, 7) display narrow violins, suggesting a concentration of cells with little to no TNFRSF4 expression.- cluster 5 has a more defined peak with a slightly higher baseline expression, suggesting more consistent expression levels within the cluster.</p></li>
</ul>
<p>This plot suggests that TNFRSF4 is highly specific to a subpopulation of cells and shows a potential inverse trend with total RNA content.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene</span> <span class="o">=</span> <span class="s2">&quot;TNFRSF4&quot;</span>
<span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">adata_pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;total_counts&quot;</span><span class="p">],</span> <span class="n">y</span><span class="o">=</span><span class="n">adata_pbmc3k</span><span class="p">[:,</span> <span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2"> Expression vs. Total RNA Counts&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/c07610ff081d14048084b415d53a80d966696062d2c5e5e393f17c7aec5ebfe2.png" src="_images/c07610ff081d14048084b415d53a80d966696062d2c5e5e393f17c7aec5ebfe2.png" />
</div>
</div>
</section>
<section id="clusterization">
<h2>Clusterization<a class="headerlink" href="#clusterization" title="Link to this heading">#</a></h2>
<p><strong>Biological Insights</strong></p>
<ul class="simple">
<li><p>The <strong>distinct clustering</strong> of neoplastic cells supports the notion of <strong>tumor heterogeneity</strong>, where multiple subpopulations of cancer cells can exist.</p></li>
<li><p>A <strong>large immune cell cluster</strong> may indicate <strong>immune infiltration</strong> within the tumor microenvironment.</p></li>
<li><p>The <strong>presence of astrocytes, OPCs, oligodendrocytes, and neurons</strong> reflects the <strong>glioblastoma‚Äôs origin</strong> in the central nervous system and the <strong>diverse cell types</strong> typically found in brain tissue.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">adata</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
    <span class="p">[</span><span class="n">adata_glioblastoma_norm</span><span class="p">,</span> <span class="n">adata_glioblastoma_raw</span><span class="p">,</span> <span class="n">adata_pbmc3k</span><span class="p">],</span> 
    <span class="p">[</span><span class="s2">&quot;Glioblastoma normalized&quot;</span><span class="p">,</span> <span class="s2">&quot;Glioblastoma Raw&quot;</span><span class="p">]</span>
<span class="p">):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>  
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>        
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">name</span><span class="p">)</span>  <span class="c1">#  UMAP</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/a67530e1fd4b3ae1274c5da8d01518ba8074ae7421ef77158e53ec3f7ac8c389.png" src="_images/a67530e1fd4b3ae1274c5da8d01518ba8074ae7421ef77158e53ec3f7ac8c389.png" />
<img alt="_images/a2aa7d3cd42c1ed4c011d265c6004927336311e252b38bcc2e424cef9010a3b3.png" src="_images/a2aa7d3cd42c1ed4c011d265c6004927336311e252b38bcc2e424cef9010a3b3.png" />
</div>
</div>
</section>
<section id="interpretation">
<h2>üîç Interpretation<a class="headerlink" href="#interpretation" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p>Each point represents an individual cell in the dataset.</p></li>
<li><p>The position in the UMAP space reflects the similarity between cells:</p>
<ul>
<li><p><strong>Nearby cells</strong> have similar gene expression profiles.</p></li>
<li><p><strong>Distant cells</strong> represent different cell populations.</p></li>
</ul>
</li>
<li><p><strong>Color indicates TNFRSF4 gene expression</strong>:</p>
<ul>
<li><p><strong>Purple (low or 0 values)</strong> ‚Üí Cells that do not express TNFRSF4.</p></li>
<li><p><strong>Green/Yellow (high values)</strong> ‚Üí Cells with higher TNFRSF4 expression.</p></li>
</ul>
</li>
<li><p><strong>Biological interpretation</strong>:</p>
<ul>
<li><p>If TNFRSF4 is expressed only in certain regions, it indicates activity in <strong>specific subpopulations</strong>.</p></li>
<li><p>If it is uniformly distributed, it suggests <strong>general expression across all cells</strong>.</p></li>
</ul>
</li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_pbmc3k</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_pbmc3k</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_pbmc3k</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_pbmc3k</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;TNFRSF4&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/a5e18d11e60f15a16ab06d225cbdbce63d040c8ecc20e2563cd2889a24aa8e05.png" src="_images/a5e18d11e60f15a16ab06d225cbdbce63d040c8ecc20e2563cd2889a24aa8e05.png" />
</div>
</div>
<p><strong>Cell Count Distribution Across Clusters</strong></p>
<p>The following table shows the number of cells in each Louvain cluster:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Cluster</p></th>
<th class="head"><p>Cell Count</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>0</strong></p></td>
<td><p><strong>1153</strong></p></td>
<td><p>Largest population, dominant cell type</p></td>
</tr>
<tr class="row-odd"><td><p><strong>1</strong></p></td>
<td><p><strong>480</strong></p></td>
<td><p>Second-largest group, possibly another major type</p></td>
</tr>
<tr class="row-even"><td><p><strong>2</strong></p></td>
<td><p><strong>341</strong></p></td>
<td><p>Medium-sized cluster, functionally distinct cells</p></td>
</tr>
<tr class="row-odd"><td><p><strong>3</strong></p></td>
<td><p><strong>303</strong></p></td>
<td><p>Similar to cluster 2, moderate presence</p></td>
</tr>
<tr class="row-even"><td><p><strong>4</strong></p></td>
<td><p><strong>157</strong></p></td>
<td><p>Smaller subpopulation, possibly a rare cell type</p></td>
</tr>
<tr class="row-odd"><td><p><strong>5</strong></p></td>
<td><p><strong>153</strong></p></td>
<td><p>Another small population, might indicate transition states</p></td>
</tr>
<tr class="row-even"><td><p><strong>6</strong></p></td>
<td><p><strong>36</strong></p></td>
<td><p>Very small cluster, could be rare or outlier cells</p></td>
</tr>
<tr class="row-odd"><td><p><strong>7</strong></p></td>
<td><p><strong>15</strong></p></td>
<td><p>Smallest group, possibly noise or rarest cell type</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="observations">
<h2><strong>Observations</strong><a class="headerlink" href="#observations" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p><strong>Cluster 0 dominates</strong>, suggesting a highly abundant cell type.</p></li>
<li><p><strong>Smaller clusters (6 &amp; 7)</strong> may represent rare populations or technical artifacts.</p></li>
<li><p>The distribution follows a common biological trend: <strong>a few large clusters and several smaller specialized groups</strong>.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;louvain&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C:\Users\henry\AppData\Local\Temp\ipykernel_4840\2660952902.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  adata_pbmc3k.obs.groupby(&quot;louvain&quot;).size()
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>louvain
0    1153
1     480
2     341
3     303
4     157
5     153
6      36
7      15
dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_pbmc3k</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;TNFRSF4&quot;</span><span class="p">,</span><span class="s2">&quot;louvain&quot;</span><span class="p">],</span> <span class="n">legend_loc</span><span class="o">=</span><span class="s2">&quot;on data&quot;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Louvain Clustering&quot;</span><span class="p">,)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: The title list is shorter than the number of panels. Using &#39;color&#39; value instead for some plots.
</pre></div>
</div>
<img alt="_images/ee775575623b9d5811ebe79fe24e87e634f58c8ba38855815adad5f5a2707545.png" src="_images/ee775575623b9d5811ebe79fe24e87e634f58c8ba38855815adad5f5a2707545.png" />
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="Datos_celulares_y_geneticos.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Exploratory Data Analysis for Glioblastoma data</p>
      </div>
    </a>
    <a class="right-next"
       href="EDA_test.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Exploratory Data Analysis for test data</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#import-data-and-information">Import Data and Information</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#adata-pbmc3k">adata_pbmc3k</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#variables-in-the-dataset">üîç Variables in the dataset</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">üîç Variables in the dataset</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#total-rna-counts-vs-mitochondrial-percentage">Total RNA Counts vs. Mitochondrial Percentage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#number-of-genes-detected-vs-mitochondrial-percentage">Number of Genes Detected vs. Mitochondrial Percentage</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#violin-plot-of-total-gene-counts-by-clusters-louvain">Violin Plot of Total Gene Counts by Clusters (louvain)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-xpression-per-cluster">Gene xpression per cluster</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#correlation-matrix">Correlation matrix</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genes-vs-mean-expression">Genes vs. Mean Expression</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#percentage-dropout-vs-mean-counts">Percentage Dropout vs. Mean Counts</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#distribution-of-more-expressed-genes-per-cluster">Distribution of more expressed genes per cluster</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#expression-of-a-specific-gene-across-clusters">Expression of a Specific Gene Across Clusters</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clusterization">Clusterization</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interpretation">üîç Interpretation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#observations"><strong>Observations</strong></a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The Jupyter Book Community
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      ¬© Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>