
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Cargue de datos &#8212; My sample book</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Datos_celulares_y_geneticos';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.png" class="logo__image only-light" alt="My sample book - Home"/>
    <script>document.write(`<img src="_static/logo.png" class="logo__image only-dark" alt="My sample book - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    RNAseq EDA
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="EDA_RNAseq.html">Cargar archivo CSV</a></li>



</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2FDatos_celulares_y_geneticos.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/Datos_celulares_y_geneticos.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Cargue de datos</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Cargue de datos</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#observar-conjunto-de-datos">Observar conjunto de datos</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#descripcion-de-las-variables-en-adata-glioblastoma-norm-obs">Descripci√≥n de las Variables en <code class="docutils literal notranslate"><span class="pre">adata_glioblastoma_norm.obs</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#significado-de-las-categorias-en-diagnosis">üìå Significado de las categor√≠as en <code class="docutils literal notranslate"><span class="pre">diagnosis</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#importancia-en-el-analisis">üìä Importancia en el an√°lisis</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#descripcion-de-las-variables">üìå Descripci√≥n de las Variables</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interpretacion">üîé Interpretaci√≥n</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#bivariate-analysis">Bivariate Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#datos-genes">Datos genes</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#descripcion-de-las-variables-en-adata-glioblastoma-norm-var">Descripci√≥n de las variables en <code class="docutils literal notranslate"><span class="pre">adata_glioblastoma_norm.var</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#analizar-la-matrix-de-expresion">Analizar la matrix de expresion</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="cargue-de-datos">
<h1>Cargue de datos<a class="headerlink" href="#cargue-de-datos" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">IPython.core.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">display</span><span class="p">,</span> <span class="n">HTML</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C:\Users\henry\AppData\Local\Temp\ipykernel_21540\1582991288.py:6: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
  from IPython.core.display import display, HTML
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_path</span> <span class="o">=</span> <span class="s2">&quot;C:/Users/henry/Desktop/python/jbook_rna/datos/Single-cell RNASeq data from Mouse Brain&quot;</span>
<span class="n">data_path2</span> <span class="o">=</span> <span class="s2">&quot;C:/Users/henry/Desktop/python/jbook_rna/datos/Theory - Intro to single-cell RNAseq Images&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">glioblastoma_norm_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="s2">&quot;glioblastoma_normalized.h5ad&quot;</span><span class="p">)</span>
<span class="n">glioblastoma_raw_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="s2">&quot;glioblastoma_raw.h5ad&quot;</span><span class="p">)</span>
<span class="n">pbmc3k_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="s2">&quot;pbmc3k.h5ad&quot;</span><span class="p">)</span>

<span class="n">adata_glioblastoma_norm</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">glioblastoma_norm_file</span><span class="p">)</span>
<span class="n">adata_glioblastoma_raw</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">glioblastoma_raw_file</span><span class="p">)</span>
<span class="n">adata_pbmc3k</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">pbmc3k_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C:\Users\henry\miniconda3\envs\ml_venv\lib\site-packages\anndata\compat\__init__.py:329: FutureWarning: Moving element from .uns[&#39;neighbors&#39;][&#39;distances&#39;] to .obsp[&#39;distances&#39;].

This is where adjacency matrices should go now.
  warn(
C:\Users\henry\miniconda3\envs\ml_venv\lib\site-packages\anndata\compat\__init__.py:329: FutureWarning: Moving element from .uns[&#39;neighbors&#39;][&#39;connectivities&#39;] to .obsp[&#39;connectivities&#39;].

This is where adjacency matrices should go now.
  warn(
</pre></div>
</div>
</div>
</div>
<section id="observar-conjunto-de-datos">
<h2>Observar conjunto de datos<a class="headerlink" href="#observar-conjunto-de-datos" title="Link to this heading">#</a></h2>
<p>Los archivos H5AD son un formato utilizado en bioinform√°tica para almacenar datos de expresi√≥n g√©nica de c√©lulas individuales en el framework Scanpy, basado en AnnData.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">display</span><span class="p">(</span><span class="n">HTML</span><span class="p">(</span><span class="s2">&quot;&lt;h3&gt;Metadatos de c√©lulas:&lt;/h3&gt;&quot;</span><span class="p">))</span>
<span class="n">display</span><span class="p">(</span><span class="n">HTML</span><span class="p">(</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">head</span><span class="p">()</span><span class="o">.</span><span class="n">to_html</span><span class="p">()))</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><h3>Metadatos de c√©lulas:</h3></div><div class="output text_html"><table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>disease_state</th>
      <th>plate_id</th>
      <th>tissue</th>
      <th>patient_id</th>
      <th>cell_type</th>
      <th>neoplastic_state</th>
      <th>diagnosis</th>
      <th>n_genes_by_counts</th>
      <th>log1p_n_genes_by_counts</th>
      <th>total_counts</th>
      <th>log1p_total_counts</th>
      <th>pct_counts_in_top_50_genes</th>
      <th>pct_counts_in_top_100_genes</th>
      <th>pct_counts_in_top_200_genes</th>
      <th>pct_counts_in_top_500_genes</th>
      <th>n_counts</th>
      <th>n_genes</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GSM2243439</th>
      <td>diagnosis: glioblastoma</td>
      <td>plate id: 1001000173</td>
      <td>tissue: Tumor</td>
      <td>patient id: BT_S2</td>
      <td>cell type: Neoplastic</td>
      <td>neoplastic: Neoplastic</td>
      <td>Neoplastic</td>
      <td>1341</td>
      <td>7.201916</td>
      <td>159262.312500</td>
      <td>11.978314</td>
      <td>65.554424</td>
      <td>76.898016</td>
      <td>89.034835</td>
      <td>98.116145</td>
      <td>159262.312500</td>
      <td>1341</td>
    </tr>
    <tr>
      <th>GSM2243444</th>
      <td>diagnosis: glioblastoma</td>
      <td>plate id: 1001000173</td>
      <td>tissue: Tumor</td>
      <td>patient id: BT_S2</td>
      <td>cell type: Neoplastic</td>
      <td>neoplastic: Neoplastic</td>
      <td>Neoplastic</td>
      <td>1456</td>
      <td>7.284135</td>
      <td>326321.875000</td>
      <td>12.695642</td>
      <td>77.530281</td>
      <td>83.443820</td>
      <td>90.526005</td>
      <td>98.158822</td>
      <td>326321.875000</td>
      <td>1456</td>
    </tr>
    <tr>
      <th>GSM2243445</th>
      <td>diagnosis: glioblastoma</td>
      <td>plate id: 1001000173</td>
      <td>tissue: Tumor</td>
      <td>patient id: BT_S2</td>
      <td>cell type: Neoplastic</td>
      <td>neoplastic: Neoplastic</td>
      <td>Neoplastic</td>
      <td>1849</td>
      <td>7.522941</td>
      <td>196996.859375</td>
      <td>12.190948</td>
      <td>50.919931</td>
      <td>59.767944</td>
      <td>70.261123</td>
      <td>85.793566</td>
      <td>196996.859375</td>
      <td>1849</td>
    </tr>
    <tr>
      <th>GSM2243452</th>
      <td>diagnosis: glioblastoma</td>
      <td>plate id: 1001000173</td>
      <td>tissue: Tumor</td>
      <td>patient id: BT_S2</td>
      <td>cell type: Neoplastic</td>
      <td>neoplastic: Neoplastic</td>
      <td>Neoplastic</td>
      <td>1788</td>
      <td>7.489412</td>
      <td>208278.578125</td>
      <td>12.246636</td>
      <td>45.863893</td>
      <td>56.620505</td>
      <td>68.793036</td>
      <td>87.116435</td>
      <td>208278.578125</td>
      <td>1788</td>
    </tr>
    <tr>
      <th>GSM2243455</th>
      <td>diagnosis: glioblastoma</td>
      <td>plate id: 1001000173</td>
      <td>tissue: Tumor</td>
      <td>patient id: BT_S2</td>
      <td>cell type: Oligodendrocyte</td>
      <td>neoplastic: Regular</td>
      <td>Oligodendrocyte</td>
      <td>2377</td>
      <td>7.774015</td>
      <td>288928.750000</td>
      <td>12.573939</td>
      <td>53.801413</td>
      <td>62.445690</td>
      <td>72.430413</td>
      <td>87.809547</td>
      <td>288928.750000</td>
      <td>2377</td>
    </tr>
  </tbody>
</table></div></div>
</div>
<p>Describir las varaobles del conjunto de datos</p>
</section>
<section id="descripcion-de-las-variables-en-adata-glioblastoma-norm-obs">
<h2>Descripci√≥n de las Variables en <code class="docutils literal notranslate"><span class="pre">adata_glioblastoma_norm.obs</span></code><a class="headerlink" href="#descripcion-de-las-variables-en-adata-glioblastoma-norm-obs" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Variable</p></th>
<th class="head"><p>Descripci√≥n</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong><code class="docutils literal notranslate"><span class="pre">disease_state</span></code></strong></p></td>
<td><p>Estado de la enfermedad del paciente (ej. sano, glioblastoma, etc.).</p></td>
</tr>
<tr class="row-odd"><td><p><strong><code class="docutils literal notranslate"><span class="pre">plate_id</span></code></strong></p></td>
<td><p>Identificador de la placa utilizada en la secuenciaci√≥n de c√©lulas.</p></td>
</tr>
<tr class="row-even"><td><p><strong><code class="docutils literal notranslate"><span class="pre">tissue</span></code></strong></p></td>
<td><p>Tipo de tejido del que se extrajo la muestra (ej. tumor, sangre, etc.).</p></td>
</tr>
<tr class="row-odd"><td><p><strong><code class="docutils literal notranslate"><span class="pre">patient_id</span></code></strong></p></td>
<td><p>Identificador √∫nico del paciente al que pertenece la muestra.</p></td>
</tr>
<tr class="row-even"><td><p><strong><code class="docutils literal notranslate"><span class="pre">cell_type</span></code></strong></p></td>
<td><p>Tipo de c√©lula identificada en la muestra (ej. Neoplastic, T cells, etc.).</p></td>
</tr>
<tr class="row-odd"><td><p><strong><code class="docutils literal notranslate"><span class="pre">neoplastic_state</span></code></strong></p></td>
<td><p>Indica si la c√©lula es neopl√°sica (cancerosa) o no.</p></td>
</tr>
<tr class="row-even"><td><p><strong><code class="docutils literal notranslate"><span class="pre">diagnosis</span></code></strong></p></td>
<td><p>Diagn√≥stico cl√≠nico asociado a la muestra (ej. glioblastoma, sano, etc.).</p></td>
</tr>
<tr class="row-odd"><td><p><strong><code class="docutils literal notranslate"><span class="pre">n_genes_by_counts</span></code></strong></p></td>
<td><p>N√∫mero de genes detectados en la c√©lula en funci√≥n de la cantidad de lecturas.</p></td>
</tr>
<tr class="row-even"><td><p><strong><code class="docutils literal notranslate"><span class="pre">log1p_n_genes_by_counts</span></code></strong></p></td>
<td><p>Logaritmo natural de <code class="docutils literal notranslate"><span class="pre">n_genes_by_counts</span></code> m√°s uno, utilizado para normalizaci√≥n.</p></td>
</tr>
<tr class="row-odd"><td><p><strong><code class="docutils literal notranslate"><span class="pre">total_counts</span></code></strong></p></td>
<td><p>N√∫mero total de lecturas (counts) asignadas a la c√©lula.</p></td>
</tr>
<tr class="row-even"><td><p><strong><code class="docutils literal notranslate"><span class="pre">log1p_total_counts</span></code></strong></p></td>
<td><p>Logaritmo natural de <code class="docutils literal notranslate"><span class="pre">total_counts</span></code> m√°s uno, para normalizaci√≥n.</p></td>
</tr>
<tr class="row-odd"><td><p><strong><code class="docutils literal notranslate"><span class="pre">pct_counts_in_top_50_genes</span></code></strong></p></td>
<td><p>Porcentaje de las cuentas totales que provienen de los 50 genes m√°s expresados.</p></td>
</tr>
<tr class="row-even"><td><p><strong><code class="docutils literal notranslate"><span class="pre">pct_counts_in_top_100_genes</span></code></strong></p></td>
<td><p>Porcentaje de las cuentas totales que provienen de los 100 genes m√°s expresados.</p></td>
</tr>
<tr class="row-odd"><td><p><strong><code class="docutils literal notranslate"><span class="pre">pct_counts_in_top_200_genes</span></code></strong></p></td>
<td><p>Porcentaje de las cuentas totales que provienen de los 200 genes m√°s expresados.</p></td>
</tr>
<tr class="row-even"><td><p><strong><code class="docutils literal notranslate"><span class="pre">pct_counts_in_top_500_genes</span></code></strong></p></td>
<td><p>Porcentaje de las cuentas totales que provienen de los 500 genes m√°s expresados.</p></td>
</tr>
<tr class="row-odd"><td><p><strong><code class="docutils literal notranslate"><span class="pre">n_counts</span></code></strong></p></td>
<td><p>N√∫mero total de cuentas en la c√©lula (puede ser redundante con <code class="docutils literal notranslate"><span class="pre">total_counts</span></code>).</p></td>
</tr>
<tr class="row-even"><td><p><strong><code class="docutils literal notranslate"><span class="pre">n_genes</span></code></strong></p></td>
<td><p>N√∫mero total de genes detectados en la c√©lula.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;disease_state&#39;,
 &#39;plate_id&#39;,
 &#39;tissue&#39;,
 &#39;patient_id&#39;,
 &#39;cell_type&#39;,
 &#39;neoplastic_state&#39;,
 &#39;diagnosis&#39;,
 &#39;n_genes_by_counts&#39;,
 &#39;log1p_n_genes_by_counts&#39;,
 &#39;total_counts&#39;,
 &#39;log1p_total_counts&#39;,
 &#39;pct_counts_in_top_50_genes&#39;,
 &#39;pct_counts_in_top_100_genes&#39;,
 &#39;pct_counts_in_top_200_genes&#39;,
 &#39;pct_counts_in_top_500_genes&#39;,
 &#39;n_counts&#39;,
 &#39;n_genes&#39;]
</pre></div>
</div>
</div>
</div>
<p>contendio de las celulas h5ad</p>
<p>numero de celulas vs nuemro de variables</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(1651, 17)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">info</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;pandas.core.frame.DataFrame&#39;&gt;
Index: 1651 entries, GSM2243439 to GSM2247071
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype   
---  ------                       --------------  -----   
 0   disease_state                1651 non-null   category
 1   plate_id                     1651 non-null   category
 2   tissue                       1651 non-null   category
 3   patient_id                   1651 non-null   category
 4   cell_type                    1651 non-null   category
 5   neoplastic_state             1651 non-null   category
 6   diagnosis                    1651 non-null   category
 7   n_genes_by_counts            1651 non-null   int64   
 8   log1p_n_genes_by_counts      1651 non-null   float64 
 9   total_counts                 1651 non-null   float32 
 10  log1p_total_counts           1651 non-null   float32 
 11  pct_counts_in_top_50_genes   1651 non-null   float64 
 12  pct_counts_in_top_100_genes  1651 non-null   float64 
 13  pct_counts_in_top_200_genes  1651 non-null   float64 
 14  pct_counts_in_top_500_genes  1651 non-null   float64 
 15  n_counts                     1651 non-null   float32 
 16  n_genes                      1651 non-null   int64   
dtypes: category(7), float32(3), float64(5), int64(2)
memory usage: 202.2+ KB
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>n_genes_by_counts</th>
      <th>log1p_n_genes_by_counts</th>
      <th>total_counts</th>
      <th>log1p_total_counts</th>
      <th>pct_counts_in_top_50_genes</th>
      <th>pct_counts_in_top_100_genes</th>
      <th>pct_counts_in_top_200_genes</th>
      <th>pct_counts_in_top_500_genes</th>
      <th>n_counts</th>
      <th>n_genes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>1651.000000</td>
      <td>1651.000000</td>
      <td>1651.000000</td>
      <td>1651.000000</td>
      <td>1651.000000</td>
      <td>1651.000000</td>
      <td>1651.000000</td>
      <td>1651.000000</td>
      <td>1651.000000</td>
      <td>1651.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>1341.784979</td>
      <td>7.177672</td>
      <td>156180.015625</td>
      <td>11.854759</td>
      <td>54.477025</td>
      <td>68.027822</td>
      <td>82.394250</td>
      <td>95.983677</td>
      <td>156180.015625</td>
      <td>1341.784979</td>
    </tr>
    <tr>
      <th>std</th>
      <td>300.774693</td>
      <td>0.223335</td>
      <td>71629.671875</td>
      <td>0.462496</td>
      <td>10.670330</td>
      <td>10.529468</td>
      <td>9.206988</td>
      <td>3.974234</td>
      <td>71629.671875</td>
      <td>300.774693</td>
    </tr>
    <tr>
      <th>min</th>
      <td>758.000000</td>
      <td>6.632002</td>
      <td>50324.871094</td>
      <td>10.826275</td>
      <td>28.037418</td>
      <td>40.053208</td>
      <td>54.781912</td>
      <td>76.990205</td>
      <td>50324.871094</td>
      <td>758.000000</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>1116.000000</td>
      <td>7.018402</td>
      <td>100596.058594</td>
      <td>11.518878</td>
      <td>46.644485</td>
      <td>60.524699</td>
      <td>75.820377</td>
      <td>94.792221</td>
      <td>100596.058594</td>
      <td>1116.000000</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>1315.000000</td>
      <td>7.182352</td>
      <td>144781.062500</td>
      <td>11.882985</td>
      <td>54.762902</td>
      <td>68.990140</td>
      <td>84.223608</td>
      <td>97.434397</td>
      <td>144781.062500</td>
      <td>1315.000000</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>1538.500000</td>
      <td>7.339213</td>
      <td>200567.851562</td>
      <td>12.208913</td>
      <td>61.652335</td>
      <td>75.890519</td>
      <td>89.952606</td>
      <td>98.702477</td>
      <td>200567.851562</td>
      <td>1538.500000</td>
    </tr>
    <tr>
      <th>max</th>
      <td>2513.000000</td>
      <td>7.829630</td>
      <td>514206.531250</td>
      <td>13.150382</td>
      <td>91.136775</td>
      <td>94.251968</td>
      <td>98.651389</td>
      <td>99.881018</td>
      <td>514206.531250</td>
      <td>2513.000000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Gr√°ficas descriptivas</p>
<p>desease_state</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;disease_state&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;diagnosis: glioblastoma&#39;]
Categories (1, object): [&#39;diagnosis: glioblastoma&#39;]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;cell_type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;cell type: Neoplastic&#39;, &#39;cell type: Oligodendrocyte&#39;, &#39;cell type: Immune cell&#39;, &#39;cell type: Vascular&#39;, &#39;cell type: OPC&#39;, &#39;cell type: Astocyte&#39;, &#39;cell type: Neuron&#39;]
Categories (7, object): [&#39;cell type: Astocyte&#39;, &#39;cell type: Immune cell&#39;, &#39;cell type: Neoplastic&#39;, &#39;cell type: Neuron&#39;, &#39;cell type: OPC&#39;, &#39;cell type: Oligodendrocyte&#39;, &#39;cell type: Vascular&#39;]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;tissue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;tissue: Tumor&#39;, &#39;tissue: Periphery&#39;]
Categories (2, object): [&#39;tissue: Periphery&#39;, &#39;tissue: Tumor&#39;]
</pre></div>
</div>
</div>
</div>
<p>‚Äòneoplastic: Neoplastic‚Äô: Indica que la c√©lula es neopl√°sica (cancerosa).
‚Äòneoplastic: Regular‚Äô: Indica que la c√©lula es normal o no cancerosa.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;neoplastic_state&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;neoplastic: Neoplastic&#39;, &#39;neoplastic: Regular&#39;]
Categories (2, object): [&#39;neoplastic: Neoplastic&#39;, &#39;neoplastic: Regular&#39;]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;diagnosis&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;Neoplastic&#39;, &#39;Oligodendrocyte&#39;, &#39;Immune cell&#39;, &#39;Vascular&#39;, &#39;OPC&#39;, &#39;Astocyte&#39;, &#39;Neuron&#39;]
Categories (7, object): [&#39;Astocyte&#39;, &#39;Immune cell&#39;, &#39;Neoplastic&#39;, &#39;Neuron&#39;, &#39;OPC&#39;, &#39;Oligodendrocyte&#39;, &#39;Vascular&#39;]
</pre></div>
</div>
</div>
</div>
</section>
<section id="significado-de-las-categorias-en-diagnosis">
<h2>üìå Significado de las categor√≠as en <code class="docutils literal notranslate"><span class="pre">diagnosis</span></code><a class="headerlink" href="#significado-de-las-categorias-en-diagnosis" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p><strong>Categor√≠a</strong></p></th>
<th class="head"><p><strong>Significado</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>Neoplastic</strong></p></td>
<td><p>C√©lulas cancerosas o tumorales en el glioblastoma. Representan c√©lulas que han sufrido alteraciones gen√©ticas y proliferan de manera descontrolada.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Oligodendrocyte</strong></p></td>
<td><p>Oligodendrocitos, c√©lulas del sistema nervioso central responsables de producir mielina, la sustancia que a√≠sla los axones neuronales.</p></td>
</tr>
<tr class="row-even"><td><p><strong>Immune cell</strong></p></td>
<td><p>C√©lulas del sistema inmune, como macr√≥fagos y linfocitos, que pueden estar infiltradas en el tumor y desempe√±ar un papel en la respuesta inmunol√≥gica contra el c√°ncer.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Vascular</strong></p></td>
<td><p>C√©lulas endoteliales y pericitos que forman los vasos sangu√≠neos en el cerebro. Su presencia en tumores puede estar relacionada con la angiog√©nesis, es decir, la formaci√≥n de nuevos vasos sangu√≠neos para nutrir el tumor.</p></td>
</tr>
<tr class="row-even"><td><p><strong>OPC (Oligodendrocyte Precursor Cells)</strong></p></td>
<td><p>C√©lulas progenitoras de oligodendrocitos, un tipo de c√©lula madre del sistema nervioso que puede diferenciarse en oligodendrocitos maduros. Se cree que pueden estar implicadas en la regeneraci√≥n celular y, en algunos casos, en la progresi√≥n tumoral.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Astrocyte</strong></p></td>
<td><p>Astrocitos, c√©lulas gliales que brindan soporte estructural y funcional a las neuronas. Pueden desempe√±ar un papel en la comunicaci√≥n neuronal y la respuesta a lesiones cerebrales.</p></td>
</tr>
<tr class="row-even"><td><p><strong>Neuron</strong></p></td>
<td><p>Neuronas, las principales c√©lulas funcionales del sistema nervioso central. Su presencia en el tumor puede deberse a la contaminaci√≥n de tejido sano en la muestra.</p></td>
</tr>
</tbody>
</table>
</div>
<section id="importancia-en-el-analisis">
<h3>üìä Importancia en el an√°lisis<a class="headerlink" href="#importancia-en-el-analisis" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>La proporci√≥n de c√©lulas <strong>neopl√°sicas</strong> puede indicar la agresividad del glioblastoma.</p></li>
<li><p>La presencia de c√©lulas del <strong>sistema inmune</strong> puede dar pistas sobre la respuesta del organismo al tumor.</p></li>
<li><p>La cantidad de c√©lulas <strong>vasculares</strong> puede sugerir qu√© tan desarrollado est√° el sistema de irrigaci√≥n del tumor.</p></li>
<li><p>Las <strong>OPC y oligodendrocitos</strong> pueden estar implicados en procesos de reparaci√≥n o diferenciaci√≥n celular en el contexto tumoral.</p></li>
</ul>
<p>‚úî <strong>Este an√°lisis es clave para entender la heterogeneidad celular en glioblastoma y su potencial respuesta a tratamientos.</strong> üöÄ</p>
<p>numero de genes por conteo</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">desc_stats</span> <span class="o">=</span> <span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;üìä Resumen Estad√≠stico de &#39;n_genes_by_counts&#39;:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">desc_stats</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>üìä Resumen Estad√≠stico de &#39;n_genes_by_counts&#39;:
 count    1651.000000
mean     1341.784979
std       300.774693
min       758.000000
25%      1116.000000
50%      1315.000000
75%      1538.500000
max      2513.000000
Name: n_genes_by_counts, dtype: float64
</pre></div>
</div>
</div>
</div>
<p>conte d egenes</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

<span class="c1"># Histograma</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Histograma de &#39;n_genes_by_counts&#39;&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;N√∫mero de genes detectados por c√©lula&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Frecuencia&quot;</span><span class="p">)</span>

<span class="c1"># Boxplot</span>
<span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Boxplot de &#39;n_genes_by_counts&#39;&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;N√∫mero de genes detectados por c√©lula&quot;</span><span class="p">)</span>

<span class="c1"># Ajustar dise√±o</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/7d15d2210945c32e79ce2cdf74b285e5e4910aeb72c0bdb49562c38cb6186dbf.png" src="_images/7d15d2210945c32e79ce2cdf74b285e5e4910aeb72c0bdb49562c38cb6186dbf.png" />
</div>
</div>
<p>log1p_numeor de genes por conyteo</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">desc_stats</span> <span class="o">=</span> <span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;log1p_n_genes_by_counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">desc_stats</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>count    1651.000000
mean        7.177672
std         0.223335
min         6.632002
25%         7.018402
50%         7.182352
75%         7.339213
max         7.829630
Name: log1p_n_genes_by_counts, dtype: float64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Crear figura con dos gr√°ficos (histograma y boxplot)</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

<span class="c1"># Histograma</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;log1p_n_genes_by_counts&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Histograma de &#39;log1p_n_genes_by_counts&#39;&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;log1p N√∫mero de genes detectados por c√©lula&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Frecuencia&quot;</span><span class="p">)</span>

<span class="c1"># Boxplot</span>
<span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;log1p_n_genes_by_counts&quot;</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Boxplot de &#39;log1p_n_genes_by_counts&#39;&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;log1p N√∫mero de genes detectados por c√©lula&quot;</span><span class="p">)</span>

<span class="c1"># Ajustar dise√±o</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/694781326dda65ca846eb7b1a8022af700ad4edfb1d05d6bfd5339a91f858847.png" src="_images/694781326dda65ca846eb7b1a8022af700ad4edfb1d05d6bfd5339a91f858847.png" />
</div>
</div>
<p>conteo porcentual de genes</p>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="descripcion-de-las-variables">
<h1>üìå Descripci√≥n de las Variables<a class="headerlink" href="#descripcion-de-las-variables" title="Link to this heading">#</a></h1>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p><strong>Variable</strong></p></th>
<th class="head"><p><strong>Descripci√≥n</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">pct_counts_in_top_50_genes</span></code></p></td>
<td><p><strong>Porcentaje de cuentas en los 50 genes m√°s abundantes</strong>. Representa el porcentaje del total de cuentas en una c√©lula que provienen de los 50 genes m√°s expresados.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">pct_counts_in_top_100_genes</span></code></p></td>
<td><p><strong>Porcentaje de cuentas en los 100 genes m√°s abundantes</strong>. Similar al anterior, pero considerando los 100 genes con mayor expresi√≥n.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">pct_counts_in_top_200_genes</span></code></p></td>
<td><p><strong>Porcentaje de cuentas en los 200 genes m√°s abundantes</strong>. Indica cu√°nto del total de la expresi√≥n celular se debe a los 200 genes m√°s expresados.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">pct_counts_in_top_500_genes</span></code></p></td>
<td><p><strong>Porcentaje de cuentas en los 500 genes m√°s abundantes</strong>. Muestra la concentraci√≥n de la expresi√≥n en los 500 genes principales.</p></td>
</tr>
</tbody>
</table>
</div>
<hr class="docutils" />
<section id="interpretacion">
<h2>üîé Interpretaci√≥n<a class="headerlink" href="#interpretacion" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p>Estas variables indican <strong>qu√© tan concentrada est√° la expresi√≥n g√©nica</strong> en un subconjunto de genes.</p></li>
<li><p>Valores altos sugieren que <strong>pocos genes dominan la expresi√≥n en una c√©lula</strong>, lo que puede indicar que ciertos genes son clave en un tipo celular espec√≠fico.</p></li>
<li><p>Valores bajos indican una <strong>distribuci√≥n m√°s uniforme</strong> de la expresi√≥n en muchos genes.</p></li>
<li><p>Pueden ayudar a detectar <strong>sesgos en la expresi√≥n</strong> o identificar tipos celulares con perfiles de expresi√≥n distintivos.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Lista de variables a analizar</span>
<span class="n">variables</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;pct_counts_in_top_50_genes&quot;</span><span class="p">,</span>
    <span class="s2">&quot;pct_counts_in_top_100_genes&quot;</span><span class="p">,</span>
    <span class="s2">&quot;pct_counts_in_top_200_genes&quot;</span><span class="p">,</span>
    <span class="s2">&quot;pct_counts_in_top_500_genes&quot;</span>
<span class="p">]</span>

<span class="c1"># Crear figura con subgr√°ficos</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">var</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">variables</span><span class="p">):</span>
    <span class="c1"># Histograma</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">var</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
    <span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Histograma de </span><span class="si">{</span><span class="n">var</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># Boxplot</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">var</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
    <span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Boxplot de </span><span class="si">{</span><span class="n">var</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Ajustar dise√±o</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/53900b61e8ef4946ff9d8e8bfa8a6257bc99dc9cf76dd9f3e9645f8f36ccb661.png" src="_images/53900b61e8ef4946ff9d8e8bfa8a6257bc99dc9cf76dd9f3e9645f8f36ccb661.png" />
</div>
</div>
<p>numero de celulas y genes</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>

<span class="c1"># Histograma de n_counts</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;n_counts&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Distribuci√≥n de n_counts&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;N√∫mero total de cuentas por c√©lula&quot;</span><span class="p">)</span>

<span class="c1"># Histograma de n_genes</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;n_genes&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Distribuci√≥n de n_genes&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;N√∫mero de genes detectados por c√©lula&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/1ada31e2a2116a4a41ffe4f73607b64c459c53be864daffe199116bae7e723b5.png" src="_images/1ada31e2a2116a4a41ffe4f73607b64c459c53be864daffe199116bae7e723b5.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>


<span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;n_counts&quot;</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Boxplot de n_counts&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;N√∫mero total de cuentas por c√©lula&quot;</span><span class="p">)</span>

<span class="c1"># Boxplot horizontal de n_genes</span>
<span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;n_genes&quot;</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Boxplot de n_genes&quot;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;N√∫mero de genes detectados por c√©lula&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/46e26fac44389e64dfd5e8efc944a8d77a12f6e86f6d91b213064c0b1a607128.png" src="_images/46e26fac44389e64dfd5e8efc944a8d77a12f6e86f6d91b213064c0b1a607128.png" />
</div>
</div>
<p>Distribuci√≥n del n√∫mero de genes por estado de enfermedad
üìå Gr√°fico sugerido: Boxplot por categor√≠a</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;disease_state&#39;</span><span class="p">],</span>
            <span class="n">y</span><span class="o">=</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes&#39;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Estado de enfermedad&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;N√∫mero de genes detectados por c√©lula&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Distribuci√≥n de &#39;n_genes_by_counts&#39; seg√∫n &#39;disease_state&#39;&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/9e2283fd39d5df74493273e12c0f866b497174fc9692b105e250f8d73e52f5f4.png" src="_images/9e2283fd39d5df74493273e12c0f866b497174fc9692b105e250f8d73e52f5f4.png" />
</div>
</div>
<section id="bivariate-analysis">
<h3>Bivariate Analysis<a class="headerlink" href="#bivariate-analysis" title="Link to this heading">#</a></h3>
<p>numerb of cells vs tissue</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tissue_counts</span> <span class="o">=</span> <span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;tissue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">tissue_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">tissue_counts</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Tissue Type&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Number of Cells&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Number of Cells per Tissue Type&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/b4c8831c6df33bd5299dba265198dda31d2de1aad2776557e001b89b82d417e0.png" src="_images/b4c8831c6df33bd5299dba265198dda31d2de1aad2776557e001b89b82d417e0.png" />
</div>
</div>
<p>number of cells for cell_type</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">cell_type_counts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">NameError</span><span class="g g-Whitespace">                                 </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">23</span><span class="p">],</span> <span class="n">line</span> <span class="mi">1</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="nb">print</span><span class="p">(</span><span class="n">cell_type_counts</span><span class="p">)</span>

<span class="ne">NameError</span>: name &#39;cell_type_counts&#39; is not defined
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_type_counts</span> <span class="o">=</span> <span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">cell_type_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">cell_type_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="n">cell_type_counts</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Number of Cells&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Number of Cells per Cell Type (Descending Order)&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>  
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C:\Users\henry\AppData\Local\Temp\ipykernel_26984\556676785.py:6: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(x=cell_type_counts.index, y=cell_type_counts.values, palette=&quot;coolwarm&quot;, order=cell_type_counts.index)
</pre></div>
</div>
<img alt="_images/26465145bc75b002af30374b8d14f3556e09d4fe517f8dae7736ae5f7225dd23.png" src="_images/26465145bc75b002af30374b8d14f3556e09d4fe517f8dae7736ae5f7225dd23.png" />
</div>
</div>
<p>Number of cells per nepplastic state</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">neoplastic_state_counts</span> <span class="o">=</span> <span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;neoplastic_state&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">neoplastic_state_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">neoplastic_state_counts</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Neoplastic State&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Number of Cells&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Number of Cells per Neoplastic State&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/ec465d610a97e332637c1aecc92ff5edef15617110108c1204f266c1ef4864ec.png" src="_images/ec465d610a97e332637c1aecc92ff5edef15617110108c1204f266c1ef4864ec.png" />
</div>
</div>
</section>
<section id="datos-genes">
<h3>Datos genes<a class="headerlink" href="#datos-genes" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">display</span><span class="p">(</span><span class="n">HTML</span><span class="p">(</span><span class="s2">&quot;&lt;h3&gt;Metadatos de genes: &lt;/h3&gt;&quot;</span><span class="p">))</span>
<span class="n">display</span><span class="p">(</span><span class="n">HTML</span><span class="p">(</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">head</span><span class="p">()</span><span class="o">.</span><span class="n">to_html</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><h3>Metadatos de genes: </h3></div><div class="output text_html"><table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>n_cells_by_counts</th>
      <th>mean_counts</th>
      <th>log1p_mean_counts</th>
      <th>pct_dropout_by_counts</th>
      <th>total_counts</th>
      <th>log1p_total_counts</th>
      <th>n_cells</th>
      <th>n_counts</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000005</th>
      <td>2</td>
      <td>0.033459</td>
      <td>0.032912</td>
      <td>99.892067</td>
      <td>62.000000</td>
      <td>4.143135</td>
      <td>2</td>
      <td>62.000000</td>
    </tr>
    <tr>
      <th>ENSG00000001561</th>
      <td>245</td>
      <td>17.581758</td>
      <td>2.922180</td>
      <td>86.778198</td>
      <td>32579.000000</td>
      <td>10.391454</td>
      <td>236</td>
      <td>32458.000000</td>
    </tr>
    <tr>
      <th>ENSG00000001629</th>
      <td>715</td>
      <td>29.176739</td>
      <td>3.407071</td>
      <td>61.413923</td>
      <td>54064.496094</td>
      <td>10.897951</td>
      <td>677</td>
      <td>52530.570312</td>
    </tr>
    <tr>
      <th>ENSG00000002587</th>
      <td>1268</td>
      <td>22.115810</td>
      <td>3.140517</td>
      <td>31.570426</td>
      <td>40980.597656</td>
      <td>10.620878</td>
      <td>1166</td>
      <td>40703.941406</td>
    </tr>
    <tr>
      <th>ENSG00000002933</th>
      <td>548</td>
      <td>45.389587</td>
      <td>3.837075</td>
      <td>70.426336</td>
      <td>84106.906250</td>
      <td>11.339856</td>
      <td>513</td>
      <td>80978.914062</td>
    </tr>
  </tbody>
</table></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(6852, 8)
</pre></div>
</div>
</div>
</div>
<p>Describir las variables</p>
</section>
</section>
<section id="descripcion-de-las-variables-en-adata-glioblastoma-norm-var">
<h2>Descripci√≥n de las variables en <code class="docutils literal notranslate"><span class="pre">adata_glioblastoma_norm.var</span></code><a class="headerlink" href="#descripcion-de-las-variables-en-adata-glioblastoma-norm-var" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">n_cells_by_counts</span></code></strong>: N√∫mero de c√©lulas en las que se detect√≥ la expresi√≥n del gen (es decir, cu√°ntas c√©lulas tienen conteos mayores a cero para este gen).</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">mean_counts</span></code></strong>: Promedio de conteos de expresi√≥n para el gen en todas las c√©lulas.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">log1p_mean_counts</span></code></strong>: Promedio de conteos de expresi√≥n despu√©s de aplicar la transformaci√≥n <code class="docutils literal notranslate"><span class="pre">log1p(x)</span> <span class="pre">=</span> <span class="pre">log(1</span> <span class="pre">+</span> <span class="pre">x)</span></code>, que reduce el efecto de valores extremos.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">pct_dropout_by_counts</span></code></strong>: Porcentaje de c√©lulas en las que el gen no fue detectado (tiene un conteo de 0). Valores altos indican genes expresados en pocas c√©lulas.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">total_counts</span></code></strong>: Suma total de conteos de expresi√≥n para ese gen en todas las c√©lulas.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">log1p_total_counts</span></code></strong>: Suma total de conteos transformada con <code class="docutils literal notranslate"><span class="pre">log1p()</span></code>, para reducir sesgos debido a genes altamente expresados.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">n_cells</span></code></strong>: N√∫mero de c√©lulas en las que el gen est√° presente (puede ser similar a <code class="docutils literal notranslate"><span class="pre">n_cells_by_counts</span></code>).</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">n_counts</span></code></strong>: Suma total de conteos de expresi√≥n para el gen, similar a <code class="docutils literal notranslate"><span class="pre">total_counts</span></code>.</p></li>
</ul>
<p>Cada fila de la tabla representa un gen identificado por su <strong>Ensembl ID</strong> (ejemplo: <code class="docutils literal notranslate"><span class="pre">ENSG00000000005</span></code>, <code class="docutils literal notranslate"><span class="pre">ENSG00000001561</span></code>), y los valores asociados indican su expresi√≥n en las c√©lulas analizadas.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;n_cells_by_counts&#39;,
 &#39;mean_counts&#39;,
 &#39;log1p_mean_counts&#39;,
 &#39;pct_dropout_by_counts&#39;,
 &#39;total_counts&#39;,
 &#39;log1p_total_counts&#39;,
 &#39;n_cells&#39;,
 &#39;n_counts&#39;]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">info</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;pandas.core.frame.DataFrame&#39;&gt;
Index: 6852 entries, ENSG00000000005 to ENSG00000284723
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   n_cells_by_counts      6852 non-null   int64  
 1   mean_counts            6852 non-null   float32
 2   log1p_mean_counts      6852 non-null   float32
 3   pct_dropout_by_counts  6852 non-null   float64
 4   total_counts           6852 non-null   float32
 5   log1p_total_counts     6852 non-null   float32
 6   n_cells                6852 non-null   int64  
 7   n_counts               6852 non-null   float32
dtypes: float32(5), float64(1), int64(2)
memory usage: 606.0+ KB
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>n_cells_by_counts</th>
      <th>mean_counts</th>
      <th>log1p_mean_counts</th>
      <th>pct_dropout_by_counts</th>
      <th>total_counts</th>
      <th>log1p_total_counts</th>
      <th>n_cells</th>
      <th>n_counts</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>6852.000000</td>
      <td>6852.000000</td>
      <td>6852.000000</td>
      <td>6852.000000</td>
      <td>6.852000e+03</td>
      <td>6852.000000</td>
      <td>6852.000000</td>
      <td>6.852000e+03</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>342.861354</td>
      <td>21.114870</td>
      <td>1.598943</td>
      <td>81.496959</td>
      <td>3.912585e+04</td>
      <td>8.164610</td>
      <td>323.185785</td>
      <td>3.763068e+04</td>
    </tr>
    <tr>
      <th>std</th>
      <td>386.570337</td>
      <td>154.856812</td>
      <td>1.432277</td>
      <td>20.861864</td>
      <td>2.869497e+05</td>
      <td>2.499461</td>
      <td>361.706252</td>
      <td>2.701092e+05</td>
    </tr>
    <tr>
      <th>min</th>
      <td>2.000000</td>
      <td>0.002698</td>
      <td>0.002695</td>
      <td>0.053967</td>
      <td>5.000000e+00</td>
      <td>1.791759</td>
      <td>2.000000</td>
      <td>5.000000e+00</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>34.000000</td>
      <td>0.330104</td>
      <td>0.285257</td>
      <td>71.397733</td>
      <td>6.116831e+02</td>
      <td>6.417847</td>
      <td>32.000000</td>
      <td>5.875100e+02</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>199.000000</td>
      <td>2.563776</td>
      <td>1.270821</td>
      <td>89.260658</td>
      <td>4.750677e+03</td>
      <td>8.466253</td>
      <td>189.000000</td>
      <td>4.593108e+03</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>530.000000</td>
      <td>13.064106</td>
      <td>2.643626</td>
      <td>98.165138</td>
      <td>2.420779e+04</td>
      <td>10.094471</td>
      <td>503.000000</td>
      <td>2.338818e+04</td>
    </tr>
    <tr>
      <th>max</th>
      <td>1852.000000</td>
      <td>7754.024902</td>
      <td>8.956097</td>
      <td>99.892067</td>
      <td>1.436821e+07</td>
      <td>16.480528</td>
      <td>1651.000000</td>
      <td>1.353124e+07</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<section id="analizar-la-matrix-de-expresion">
<h3>Analizar la matrix de expresion<a class="headerlink" href="#analizar-la-matrix-de-expresion" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">X</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[0.       , 0.       , 0.       , ..., 0.       , 0.       ,
        0.       ],
       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,
        0.       ],
       [0.       , 0.       , 5.6354756, ..., 0.       , 0.       ,
        0.       ],
       ...,
       [0.       , 0.       , 6.7574944, ..., 0.       , 0.       ,
        0.       ],
       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,
        0.       ],
       [0.       , 0.       , 5.4751015, ..., 0.       , 0.       ,
        0.       ]], dtype=float32)
</pre></div>
</div>
</div>
</div>
<p>Ver qu√© genes corresponden a cada columna de X
python
Copiar
Editar</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;ENSG00000000005&#39;, &#39;ENSG00000001561&#39;, &#39;ENSG00000001629&#39;,
       &#39;ENSG00000002587&#39;, &#39;ENSG00000002933&#39;, &#39;ENSG00000003056&#39;,
       &#39;ENSG00000003137&#39;, &#39;ENSG00000004478&#39;, &#39;ENSG00000004779&#39;,
       &#39;ENSG00000004799&#39;,
       ...
       &#39;ENSG00000283945&#39;, &#39;ENSG00000283980&#39;, &#39;ENSG00000284308&#39;,
       &#39;ENSG00000284337&#39;, &#39;ENSG00000284413&#39;, &#39;ENSG00000284516&#39;,
       &#39;ENSG00000284573&#39;, &#39;ENSG00000284680&#39;, &#39;ENSG00000284690&#39;,
       &#39;ENSG00000284723&#39;],
      dtype=&#39;object&#39;, name=&#39;index&#39;, length=6852)
</pre></div>
</div>
</div>
</div>
<p>Ver qu√© c√©lulas corresponden a cada fila de X</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;GSM2243439&#39;, &#39;GSM2243444&#39;, &#39;GSM2243445&#39;, &#39;GSM2243452&#39;, &#39;GSM2243455&#39;,
       &#39;GSM2243456&#39;, &#39;GSM2243457&#39;, &#39;GSM2243459&#39;, &#39;GSM2243460&#39;, &#39;GSM2243464&#39;,
       ...
       &#39;GSM2247059&#39;, &#39;GSM2247062&#39;, &#39;GSM2247063&#39;, &#39;GSM2247064&#39;, &#39;GSM2247065&#39;,
       &#39;GSM2247066&#39;, &#39;GSM2247067&#39;, &#39;GSM2247068&#39;, &#39;GSM2247070&#39;, &#39;GSM2247071&#39;],
      dtype=&#39;object&#39;, name=&#39;index&#39;, length=1651)
</pre></div>
</div>
</div>
</div>
<p>accederca ka info</p>
<p>por gen</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_id</span> <span class="o">=</span> <span class="s2">&quot;ENSG00000001561&quot;</span>  <span class="c1"># Cambia esto por el ID del gen que te interesa</span>
<span class="n">adata_glioblastoma_norm</span><span class="p">[:,</span> <span class="n">gene_id</span><span class="p">]</span><span class="o">.</span><span class="n">X</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>ArrayView([[0.],
           [0.],
           [0.],
           ...,
           [0.],
           [0.],
           [0.]], dtype=float32)
</pre></div>
</div>
</div>
</div>
<p>por celula</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_id</span> <span class="o">=</span> <span class="s2">&quot;GSM2243439&quot;</span>  <span class="c1"># Cambia esto por el ID de la c√©lula que te interesa</span>
<span class="n">adata_glioblastoma_norm</span><span class="p">[</span><span class="n">cell_id</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">X</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>ArrayView([[0., 0., 0., ..., 0., 0., 0.]], dtype=float32)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filtrar genes expresados en pocas c√©lulas</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata_glioblastoma_norm</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filtrar c√©lulas con pocos genes detectados</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata_glioblastoma_norm</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>50 celulas tinee 30000 genes</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;total_counts&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Total de cuentas por c√©lula&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Frecuencia&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/710bc19fcf269efc80a9b92e936a66720d55536093cc4fd5e4b8265abacff947.png" src="_images/710bc19fcf269efc80a9b92e936a66720d55536093cc4fd5e4b8265abacff947.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">adata_glioblastoma_norm</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;total_counts&quot;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/10f04febc86c53b55de39885375c6bee0f477c28f6f59e8c1c1cdf689262a556.png" src="_images/10f04febc86c53b55de39885375c6bee0f477c28f6f59e8c1c1cdf689262a556.png" />
</div>
</div>
<p>pbmc3k.h5ad</p>
<p>Conjunto de datos de 3,000 c√©lulas mononucleares de sangre perif√©rica (PBMCs).
Se usa como referencia en an√°lisis de expresi√≥n g√©nica.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">adata</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
    <span class="p">[</span><span class="n">adata_glioblastoma_norm</span><span class="p">,</span> <span class="n">adata_glioblastoma_raw</span><span class="p">,</span> <span class="n">adata_pbmc3k</span><span class="p">],</span> 
    <span class="p">[</span><span class="s2">&quot;Glioblastoma Normalizado&quot;</span><span class="p">,</span> <span class="s2">&quot;Glioblastoma Raw&quot;</span><span class="p">,</span> <span class="s2">&quot;PBMC3K&quot;</span><span class="p">]</span>
<span class="p">):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>   <span class="c1"># Construir la gr√°fica de vecinos</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>        <span class="c1"># Calcular la reducci√≥n de dimensionalidad</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">name</span><span class="p">)</span>  <span class="c1"># Graficar UMAP</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/978c2f6054ca4012aea4f18f6acde91465e69d0e838ac4e50eff11a1e448e3bc.png" src="_images/978c2f6054ca4012aea4f18f6acde91465e69d0e838ac4e50eff11a1e448e3bc.png" />
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>c:\Users\henry\miniconda3\envs\ml_venv\lib\site-packages\scanpy\tools\_utils.py:41: UserWarning: You‚Äôre trying to run this on 7683 dimensions of `.X`, if you really want this, set `use_rep=&#39;X&#39;`.
         Falling back to preprocessing with `sc.pp.pca` and default params.
  warnings.warn(
</pre></div>
</div>
<img alt="_images/a2aa7d3cd42c1ed4c011d265c6004927336311e252b38bcc2e424cef9010a3b3.png" src="_images/a2aa7d3cd42c1ed4c011d265c6004927336311e252b38bcc2e424cef9010a3b3.png" />
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">KeyError</span><span class="g g-Whitespace">                                  </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">11</span><span class="p">],</span> <span class="n">line</span> <span class="mi">7</span>
<span class="g g-Whitespace">      </span><span class="mi">5</span> <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>   <span class="c1"># Construir la gr√°fica de vecinos</span>
<span class="g g-Whitespace">      </span><span class="mi">6</span> <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>        <span class="c1"># Calcular la reducci√≥n de dimensionalidad</span>
<span class="ne">----&gt; </span><span class="mi">7</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">name</span><span class="p">)</span>  <span class="c1"># Graficar UMAP</span>

<span class="nn">File c:\Users\henry\miniconda3\envs\ml_venv\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:691,</span> in <span class="ni">umap</span><span class="nt">(adata, **kwargs)</span>
<span class="g g-Whitespace">    </span><span class="mi">632</span> <span class="nd">@_wraps_plot_scatter</span>
<span class="g g-Whitespace">    </span><span class="mi">633</span> <span class="nd">@_doc_params</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">634</span>     <span class="n">adata_color_etc</span><span class="o">=</span><span class="n">doc_adata_color_etc</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">638</span> <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">639</span> <span class="k">def</span><span class="w"> </span><span class="nf">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Figure</span> <span class="o">|</span> <span class="n">Axes</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="n">Axes</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">640</span><span class="w">     </span><span class="sd">&quot;&quot;&quot;\</span>
<span class="g g-Whitespace">    </span><span class="mi">641</span><span class="sd">     Scatter plot in UMAP basis.</span>
<span class="g g-Whitespace">    </span><span class="mi">642</span><span class="sd"> </span>
<span class="sd">   (...)</span>
<span class="g g-Whitespace">    </span><span class="mi">689</span><span class="sd">     tl.umap</span>
<span class="g g-Whitespace">    </span><span class="mi">690</span><span class="sd">     &quot;&quot;&quot;</span>
<span class="ne">--&gt; </span><span class="mi">691</span>     <span class="k">return</span> <span class="n">embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;umap&quot;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

<span class="nn">File c:\Users\henry\miniconda3\envs\ml_venv\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:279,</span> in <span class="ni">embedding</span><span class="nt">(adata, basis, color, mask_obs, gene_symbols, use_raw, sort_order, edges, edges_width, edges_color, neighbors_key, arrows, arrows_kwds, groups, components, dimensions, layer, projection, scale_factor, color_map, cmap, palette, na_color, na_in_legend, size, frameon, legend_fontsize, legend_fontweight, legend_loc, legend_fontoutline, colorbar_loc, vmax, vmin, vcenter, norm, add_outline, outline_width, outline_color, ncols, hspace, wspace, title, show, save, ax, return_fig, marker, **kwargs)</span>
<span class="g g-Whitespace">    </span><span class="mi">277</span> <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="p">(</span><span class="n">value_to_plot</span><span class="p">,</span> <span class="n">dims</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">color</span><span class="p">,</span> <span class="n">dimensions</span><span class="p">)):</span>
<span class="g g-Whitespace">    </span><span class="mi">278</span>     <span class="n">kwargs_scatter</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>  <span class="c1"># is potentially mutated for each plot</span>
<span class="ne">--&gt; </span><span class="mi">279</span>     <span class="n">color_source_vector</span> <span class="o">=</span> <span class="n">_get_color_source_vector</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">280</span>         <span class="n">adata</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">281</span>         <span class="n">value_to_plot</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">282</span>         <span class="n">layer</span><span class="o">=</span><span class="n">layer</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">283</span>         <span class="n">mask_obs</span><span class="o">=</span><span class="n">mask_obs</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">284</span>         <span class="n">use_raw</span><span class="o">=</span><span class="n">use_raw</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">285</span>         <span class="n">gene_symbols</span><span class="o">=</span><span class="n">gene_symbols</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">286</span>         <span class="n">groups</span><span class="o">=</span><span class="n">groups</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">287</span>     <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">288</span>     <span class="n">color_vector</span><span class="p">,</span> <span class="n">color_type</span> <span class="o">=</span> <span class="n">_color_vector</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">289</span>         <span class="n">adata</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">290</span>         <span class="n">value_to_plot</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">293</span>         <span class="n">na_color</span><span class="o">=</span><span class="n">na_color</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">294</span>     <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">296</span>     <span class="c1"># Order points</span>

<span class="nn">File c:\Users\henry\miniconda3\envs\ml_venv\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:1207,</span> in <span class="ni">_get_color_source_vector</span><span class="nt">(adata, value_to_plot, mask_obs, use_raw, gene_symbols, layer, groups)</span>
<span class="g g-Whitespace">   </span><span class="mi">1205</span>     <span class="n">values</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">obs_vector</span><span class="p">(</span><span class="n">value_to_plot</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1206</span> <span class="k">else</span><span class="p">:</span>
<span class="ne">-&gt; </span><span class="mi">1207</span>     <span class="n">values</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs_vector</span><span class="p">(</span><span class="n">value_to_plot</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="n">layer</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1208</span> <span class="k">if</span> <span class="n">mask_obs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1209</span>     <span class="n">values</span><span class="p">[</span><span class="o">~</span><span class="n">mask_obs</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

<span class="nn">File c:\Users\henry\miniconda3\envs\ml_venv\lib\site-packages\anndata\_core\anndata.py:1301,</span> in <span class="ni">AnnData.obs_vector</span><span class="nt">(self, k, layer)</span>
<span class="g g-Whitespace">   </span><span class="mi">1295</span>         <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1296</span>             <span class="s2">&quot;In a future version of AnnData, access to `.X` by passing&quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1297</span>             <span class="s2">&quot; `layer=&#39;X&#39;` will be removed. Instead pass `layer=None`.&quot;</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1298</span>             <span class="ne">FutureWarning</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1299</span>         <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1300</span>         <span class="n">layer</span> <span class="o">=</span> <span class="kc">None</span>
<span class="ne">-&gt; </span><span class="mi">1301</span> <span class="k">return</span> <span class="n">get_vector</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="s2">&quot;obs&quot;</span><span class="p">,</span> <span class="s2">&quot;var&quot;</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="n">layer</span><span class="p">)</span>

<span class="nn">File c:\Users\henry\miniconda3\envs\ml_venv\lib\site-packages\anndata\_core\index.py:221,</span> in <span class="ni">get_vector</span><span class="nt">(adata, k, coldim, idxdim, layer)</span>
<span class="g g-Whitespace">    </span><span class="mi">217</span>     <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">218</span>         <span class="sa">f</span><span class="s2">&quot;Key </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2"> could be found in both .</span><span class="si">{</span><span class="n">idxdim</span><span class="si">}</span><span class="s2">_names and .</span><span class="si">{</span><span class="n">coldim</span><span class="si">}</span><span class="s2">.columns&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">219</span>     <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">220</span> <span class="k">elif</span> <span class="p">(</span><span class="n">in_col</span> <span class="o">+</span> <span class="n">in_idx</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">221</span>     <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">222</span>         <span class="sa">f</span><span class="s2">&quot;Could not find key </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2"> in .</span><span class="si">{</span><span class="n">idxdim</span><span class="si">}</span><span class="s2">_names or .</span><span class="si">{</span><span class="n">coldim</span><span class="si">}</span><span class="s2">.columns.&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">223</span>     <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">224</span> <span class="k">elif</span> <span class="n">in_col</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">225</span>     <span class="k">return</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">coldim</span><span class="p">)[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

<span class="ne">KeyError</span>: &#39;Could not find key cell_type in .var_names or .obs.columns.&#39;
</pre></div>
</div>
<img alt="_images/1666bd7d93a7212337303e4225edbe0a71238a8ea1792f9b9a74430b38f28f7e.png" src="_images/1666bd7d93a7212337303e4225edbe0a71238a8ea1792f9b9a74430b38f28f7e.png" />
</div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Cargue de datos</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#observar-conjunto-de-datos">Observar conjunto de datos</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#descripcion-de-las-variables-en-adata-glioblastoma-norm-obs">Descripci√≥n de las Variables en <code class="docutils literal notranslate"><span class="pre">adata_glioblastoma_norm.obs</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#significado-de-las-categorias-en-diagnosis">üìå Significado de las categor√≠as en <code class="docutils literal notranslate"><span class="pre">diagnosis</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#importancia-en-el-analisis">üìä Importancia en el an√°lisis</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#descripcion-de-las-variables">üìå Descripci√≥n de las Variables</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interpretacion">üîé Interpretaci√≥n</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#bivariate-analysis">Bivariate Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#datos-genes">Datos genes</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#descripcion-de-las-variables-en-adata-glioblastoma-norm-var">Descripci√≥n de las variables en <code class="docutils literal notranslate"><span class="pre">adata_glioblastoma_norm.var</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#analizar-la-matrix-de-expresion">Analizar la matrix de expresion</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The Jupyter Book Community
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      ¬© Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>